// @bmw-ds/components/esm2015/inputfield/bmw-input.directive.js
var BmwInputDirective_1;
import { __decorate, __param } from "tslib";
import { InputfieldComponent } from './inputfield.component';
import { BaseInput } from './base-input';
import { Directive, HostBinding, Input, ElementRef, AfterViewInit, Output, EventEmitter, Host, Optional } from '@angular/core';
const SUPPORTED_TYPES = ['text', 'file', 'number', 'password'];
let BmwInputDirective = BmwInputDirective_1 = class BmwInputDirective {
    constructor(_el, _inputField) {
        this._el = _el;
        this._inputField = _inputField;
        this.classBmwInput = true;
        this._type = 'text';
        this.disableToggle = new EventEmitter();
    }
    // when we bind to attr.disabled we need to set it to 'disabled'
    // or remove it with null because disabled="false" is not the right way
    // to enable the input again, especially the :disabled pseudo selector in css
    // would still be in effect
    get isDisabled() {
        return this.disabled ? 'disabled' : null;
    }
    set disabled(value) {
        this._disabled = value;
        this.disableToggle.emit(value);
    }
    get disabled() {
        return this._inputField && this._inputField.disabled ? true : this._disabled;
    }
    set withIcon(value) {
        if (value) {
            this._el.nativeElement.classList.add('with-icon');
        }
        else {
            this._el.nativeElement.classList.remove('with-icon');
        }
    }
    get id() {
        return this._id;
    }
    set id(value) {
        this._id = value;
    }
    get type() {
        return this._type;
    }
    set type(value) {
        if (SUPPORTED_TYPES.indexOf(value) === -1) {
            throw new Error(`Invalid or unsupported type: ${value}`);
        }
        this._type = value;
    }
    ngAfterViewInit() {
        const list = Array.from(this._el.nativeElement.classList);
        this.placeholder = this._el.nativeElement.placeholder;
        Promise.resolve().then(() => {
            this.classes = list
                .map((item) => {
                if (item.match(/bmw__grid__cell.*/)) {
                    return item;
                }
                return null;
            })
                .filter(item => item);
        });
    }
};
BmwInputDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: InputfieldComponent, decorators: [{ type: Optional }, { type: Host }] }
];
__decorate([
    HostBinding('class.bmw-input')
], BmwInputDirective.prototype, "classBmwInput", void 0);
__decorate([
    HostBinding('attr.disabled')
], BmwInputDirective.prototype, "isDisabled", null);
__decorate([
    Input()
], BmwInputDirective.prototype, "disabled", null);
__decorate([
    HostBinding('tabindex')
], BmwInputDirective.prototype, "tabIndex", void 0);
__decorate([
    Output()
], BmwInputDirective.prototype, "disableToggle", void 0);
__decorate([
    HostBinding('attr.id'),
    Input()
], BmwInputDirective.prototype, "id", null);
__decorate([
    Input()
], BmwInputDirective.prototype, "type", null);
BmwInputDirective = BmwInputDirective_1 = __decorate([
    Directive({
        selector: 'input[bmwInput], textarea[bmwInput]',
        providers: [
            {
                provide: BaseInput,
                useExisting: BmwInputDirective_1
            }
        ]
    }),
    __param(1, Optional()), __param(1, Host())
], BmwInputDirective);
export { BmwInputDirective };
