// @bmw-ds/components/esm5/dropdown/dropdown.component.js
import { __decorate } from "tslib";
import { AfterContentInit, Component, ContentChildren, EventEmitter, forwardRef, HostBinding, HostListener, Input, OnDestroy, OnInit, Output, QueryList, ElementRef, isDevMode } from '@angular/core';
import { DropdownItemComponent } from './dropdown-item.component';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var DropdownComponent = /** @class */ (function () {
    function DropdownComponent(_el) {
        this._el = _el;
        this.valueChange = new EventEmitter();
        this.opened = false;
        // this is emitted in ngOnDestroy
        this._destroy = new Subject();
        this._onChange = function () { };
        this._onTouched = function () { };
    }
    DropdownComponent_1 = DropdownComponent;
    Object.defineProperty(DropdownComponent.prototype, "value", {
        get: function () {
            return this._value;
        },
        set: function (newValue) {
            var _this = this;
            if (newValue !== this._value) {
                this._value = newValue;
                this._onChange(newValue);
                if (this.childDropdownMenuItems) {
                    this.childDropdownMenuItems.forEach(function (item) {
                        if (_this._value === item.value) {
                            item.selected = true;
                            _this.label = item.label;
                            _this.selectedItem = item;
                        }
                        else {
                            item.selected = false;
                        }
                    });
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    DropdownComponent.prototype.ngOnInit = function () {
        if (isDevMode()) {
            console.warn('Warning - <bmw-dropdown> component is deprecated and will be removed in the near future.');
            console.warn('Please use <bmw-combobox> instead https://designsystem.bmwgroup.net/components/form-combo-box');
        }
        if (!this.id) {
            this.id = Math.random()
                .toString(36)
                .substring(2);
        }
    };
    DropdownComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        // do this in the next cycle to avoid issues with dynamic component creation
        // TODO refactor this hack: call change detection when needed
        setTimeout(function () {
            _this._prepareItems();
            _this._validate();
            _this.childDropdownMenuItems.changes.pipe(takeUntil(_this._destroy)).subscribe(function () {
                _this._prepareItems();
            });
            _this._setGridClasses();
        });
    };
    DropdownComponent.prototype.ngOnDestroy = function () {
        this._destroy.next();
        this._destroy.complete();
    };
    DropdownComponent.prototype.onKeydown = function (event) {
        var keyName = event.key;
        event.preventDefault();
        if (keyName.slice(-4) === 'Down') {
            this.toggle(true);
            if (this._hovered === this.childDropdownMenuItems.length - 1)
                return;
            typeof this._hovered !== 'undefined' ? this._hovered++ : (this._hovered = 0);
            this._hover(this._hovered);
        }
        else if (keyName.slice(-2) === 'Up') {
            this.toggle(true);
            if (this._hovered === 0)
                return;
            typeof this._hovered !== 'undefined' ? this._hovered-- : (this._hovered = this.childDropdownMenuItems.length - 1);
            this._hover(this._hovered);
        }
        else if (keyName === 'Enter') {
            var item = this.childDropdownMenuItems[this._hovered];
            if (!item.disabled) {
                this._select(item.value);
                this.toggle(false);
            }
        }
        else if (keyName.slice(0, 3) === 'Esc') {
            this.toggle(false);
        }
    };
    DropdownComponent.prototype.onClickOutside = function ($e) {
        if ($e && !this._el.nativeElement.contains($e)) {
            this.toggle(false);
        }
    };
    DropdownComponent.prototype._prepareItems = function () {
        var _this = this;
        this.childDropdownMenuItems.forEach(function (item) {
            if (_this._value === item.value) {
                _this._select(item.value);
            }
            if (item.clicked.observers.length > 0) {
                return;
            }
            item.clicked.pipe(takeUntil(_this._destroy)).subscribe(function (selectedItem) {
                _this._select(selectedItem.value);
            });
        });
    };
    DropdownComponent.prototype.click = function (event) {
        this.toggle();
    };
    DropdownComponent.prototype.blur = function () {
        this._onTouched();
    };
    DropdownComponent.prototype.toggle = function (state) {
        if (typeof state !== 'undefined') {
            this.opened = state;
        }
        else {
            this.opened = !this.opened;
        }
    };
    DropdownComponent.prototype._validate = function () {
        var _this = this;
        if (typeof this.valid !== 'undefined' && this.valid !== null) {
            this.isValid = this.valid;
        }
        else {
            setTimeout(function () {
                var invalid = Array.from(_this._el.nativeElement.classList).includes('ng-invalid');
                _this.isValid = (_this.required !== true || _this._value) && !invalid;
            });
        }
    };
    DropdownComponent.prototype._select = function (value) {
        var _this = this;
        this.value = value;
        this.childDropdownMenuItems.forEach(function (item) {
            if (item.value === value) {
                item.selected = true;
                _this.label = item.label;
                _this.selectedItem = item;
            }
            else {
                item.selected = false;
            }
        });
        this.valueChange.emit(value);
        this._validate();
    };
    DropdownComponent.prototype._hover = function (selectedItemIndex) {
        this.childDropdownMenuItems.forEach(function (item, index) {
            if (index === selectedItemIndex) {
                item.keyboardFocus = true;
            }
            else {
                item.keyboardFocus = false;
            }
        });
    };
    DropdownComponent.prototype._setGridClasses = function () {
        var _this = this;
        var listOfClasses = Array.from(this._el.nativeElement.classList);
        this.gridClasses = listOfClasses.filter(function (item) { return item.match(/bmw__grid__cell--.*/); });
        if (this.gridClasses.length) {
            var wrapperClassList_1 = this._el.nativeElement.querySelector('div.dropdown-outer-wrapper').classList;
            wrapperClassList_1.value.split(' ').forEach(function (className) {
                if (className.match(/bmw__grid__cell--.*/)) {
                    wrapperClassList_1.remove(className);
                }
            });
            this.gridClasses.forEach(function (className) {
                _this._el.nativeElement.querySelector('div.dropdown-outer-wrapper').classList.add(className);
                _this._el.nativeElement.classList.remove(className);
            });
        }
    };
    DropdownComponent.prototype.registerOnChange = function (fn) {
        this._onChange = fn;
    };
    DropdownComponent.prototype.registerOnTouched = function (fn) {
        this._onTouched = fn;
    };
    DropdownComponent.prototype.setDisabledState = function (isDisabled) {
        this.disabled = isDisabled;
    };
    DropdownComponent.prototype.writeValue = function (value) {
        this.value = value;
    };
    var DropdownComponent_1;
    DropdownComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        ContentChildren(DropdownItemComponent)
    ], DropdownComponent.prototype, "childDropdownMenuItems", void 0);
    __decorate([
        Input()
    ], DropdownComponent.prototype, "disabled", void 0);
    __decorate([
        Input()
    ], DropdownComponent.prototype, "required", void 0);
    __decorate([
        Input()
    ], DropdownComponent.prototype, "valid", void 0);
    __decorate([
        HostBinding('attr.id'),
        Input()
    ], DropdownComponent.prototype, "id", void 0);
    __decorate([
        Output()
    ], DropdownComponent.prototype, "valueChange", void 0);
    __decorate([
        Input()
    ], DropdownComponent.prototype, "value", null);
    __decorate([
        HostListener('keydown', ['$event'])
    ], DropdownComponent.prototype, "onKeydown", null);
    __decorate([
        HostListener('document:click', ['$event.target'])
    ], DropdownComponent.prototype, "onClickOutside", null);
    DropdownComponent = DropdownComponent_1 = __decorate([
        Component({
            selector: 'bmw-dropdown',
            template: "<div class=\"dropdown-wrapper\" [ngClass]=\"{ bmw__grid__inner: gridClasses?.length, disabled: disabled }\">\n  <ng-content select=\"bmw-label\"></ng-content>\n  <div class=\"dropdown-outer-wrapper bmw__grid__cell bmw__grid__cell--span-2 bmw__grid__cell--span-4-mobile\">\n    <div tabindex=\"0\" (click)=\"click($event)\" (blur)=\"blur()\" class=\"wrapper\" [class.invalid]=\"!isValid\">\n      <div class=\"select\" [class.open]=\"opened\">\n        <span *ngIf=\"selectedItem\" class=\"[placeholder]\"> {{ selectedItem.label }}</span>\n\n        <span *ngIf=\"!selectedItem\" class=\"placeholder\"> <ng-content select=\"label\"></ng-content> </span>\n      </div>\n      <div [class.open]=\"opened\" class=\"dropdown-menu\">\n        <ng-content select=\"bmw-dropdown-item\"></ng-content>\n      </div>\n    </div>\n    <ng-content select=\"bmw-error-message\"></ng-content>\n  </div>\n</div>\n",
            providers: [
                {
                    provide: NG_VALUE_ACCESSOR,
                    multi: true,
                    useExisting: forwardRef(function () { return DropdownComponent_1; })
                }
            ],
            styles: [".bmw__grid__inner.disabled{cursor:no-drop}.bmw__grid__inner.disabled div.wrapper{border:none;pointer-events:none}.bmw__grid__inner.disabled div.wrapper .select{border-bottom:1px solid var(--dropdown__disabled__color)}.bmw__grid__inner.disabled div.wrapper span.placeholder{color:var(--dropdown__disabled__color)!important}.bmw__grid__inner.disabled ::ng-deep label{cursor:no-drop;color:var(--inputfield__disabled__color)}:host:not(.ignore-invalid) .bmw__grid__inner:not(.disabled){grid-row-gap:0}:host:not(.ignore-invalid) .bmw__grid__inner:not(.disabled) div.wrapper.invalid{border:none}:host:not(.ignore-invalid) .bmw__grid__inner:not(.disabled) div.wrapper.invalid .select{border-bottom:1px solid!important;background-color:var(--dropdown__error__background-color)!important}div.wrapper{border:none;position:relative;height:var(--inputfield__default__height);cursor:pointer}div.wrapper:focus,div.wrapper:hover{outline:0}div.wrapper .select{width:100%;position:relative;display:block;border-bottom:1px solid var(--dropdown__default__border-bottom-color);padding-left:var(--dropdown__default__padding-left);padding-right:var(--dropdown__default__padding-right);box-sizing:border-box;background-color:var(--dropdown__default__background-color)}div.wrapper .select:hover{border-bottom-color:var(--dropdown__focus__border-bottom-color)}::ng-deep .reversed div.wrapper .select{background-color:var(--dropdown__reversed__background-color)!important}div.wrapper .select span{display:block;text-overflow:ellipsis;white-space:nowrap;overflow:hidden;height:var(--inputfield__default__height);width:100%;min-width:100%;line-height:var(--inputfield__default__height);font-size:var(--dropdown__default__font-size);font-family:Arial;color:var(--inputfield__default__color)}div.wrapper .select span.placeholder{color:var(--dropdown__placeholder__color)}div.wrapper .select span::after{content:'\\e92f';font-family:iwp;position:absolute;right:calc(var(--dropdown__arrow__line-height)/ 2 - var(--dropdown__arrow__font-size)/ 2 + var(--dropdown__arrow__padding-right));font-size:var(--dropdown__arrow__font-size);line-height:var(--dropdown__arrow__line-height);top:calc(50% - var(--dropdown__arrow__line-height)/ 2 - 1px)}div.wrapper .select.open span::after{content:'\\e932'}:host{position:relative}:host ::ng-deep label{cursor:inherit}div.dropdown-menu{display:none;position:absolute;z-index:var(--dropdown__menu__z-index);overflow-y:auto;width:calc(100% - 2px);max-height:calc(5.66 * (var(--inputfield__default__height)));border:solid 1px var(--dropdown__dropdown-menu__border-color);border-top:none;top:calc(var(--inputfield__default__height) + 1px)}div.dropdown-menu.open{display:block}.show-error :host.ng-invalid.ng-dirty div.wrapper .select{border-bottom:1px solid var(--dropdown__error__color);background-color:var(--dropdown__error__background-color)!important}"]
        })
    ], DropdownComponent);
    return DropdownComponent;
}());
export { DropdownComponent };
