// @bmw-ds/components/esm5/navigation-menu/navigation-menu-submenu.component.js
import { __decorate } from "tslib";
import { Component, Input } from '@angular/core';
var NavigationMenuSubmenuComponent = /** @class */ (function () {
    function NavigationMenuSubmenuComponent() {
        this.opened = false;
    }
    NavigationMenuSubmenuComponent.prototype.toggleMenu = function ($event) {
        $event.stopPropagation();
        this.opened = !this.opened;
    };
    NavigationMenuSubmenuComponent.prototype.ngOnInit = function () {
        this.opened = this.label ? false : true;
    };
    __decorate([
        Input()
    ], NavigationMenuSubmenuComponent.prototype, "label", void 0);
    NavigationMenuSubmenuComponent = __decorate([
        Component({
            selector: 'bmw-navigation-menu-submenu',
            template: "<div class=\"navigation-submenu-wrapper\" [ngClass]=\"{ 'second-level': label }\">\n  <div class=\"submenu-label\" (click)=\"toggleMenu($event)\" *ngIf=\"label\">\n    <span>{{ label }}</span>\n    <i class=\"chevron\" [ngClass]=\"opened ? 'iwp-icon-gen_nav_arrow_up' : 'iwp-icon-gen_nav_arrow_down'\"></i>\n  </div>\n  <div class=\"submenu-items\" *ngIf=\"opened\">\n    <ng-content></ng-content>\n  </div>\n</div>\n",
            styles: [".submenu-label{background-color:var(--navigation-menu__default__submenu__background-color);cursor:pointer;display:flex;align-items:center;justify-content:space-between;padding:calc(var(--navigation-menu__icon__padding) * 2) 0;padding-right:calc(var(--navigation-menu__icon__padding) * 2);padding-left:calc(var(--navigation-menu__icon__padding) * 2 + var(--navigation-menu__icon__font-size) + var(--navigation-menu__item__padding-left) + var(--navigation-menu__submenu-item__padding-left))}.submenu-label span{text-overflow:ellipsis;overflow:hidden;white-space:nowrap}.submenu-label:hover{background-color:var(--navigation-menu__hover__background-color)}.submenu-label:focus{background-color:var(--navigation-menu__hover__background-color);outline:solid 1px;outline-color:var(--color-bmw-highlight);outline-offset:-1px}:before{left:0}::ng-deep .navigation-submenu-wrapper.second-level .submenu-item-wrapper{padding-left:calc(var(--navigation-menu__icon__padding) * 2 + var(--navigation-menu__icon__font-size) + var(--navigation-menu__item__padding-left) + var(--navigation-menu__submenu-item__padding-left) + var(--navigation-menu__submenu-item__second-level__padding-left))}"]
        })
    ], NavigationMenuSubmenuComponent);
    return NavigationMenuSubmenuComponent;
}());
export { NavigationMenuSubmenuComponent };
