// @bmw-ds/components/esm5/navigation-menu/navigation-menu.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NavigationMenuComponent } from './navigation-menu.component';
import { NavigationMenuItemComponent } from './navigation-menu-item.component';
import { NavigationMenuSubmenuItemComponent } from './navigation-menu-submenu-item.component';
import { NavigationMenuSubmenuComponent } from './navigation-menu-submenu.component';
var NavigationMenuModule = /** @class */ (function () {
    function NavigationMenuModule() {
    }
    NavigationMenuModule = __decorate([
        NgModule({
            imports: [CommonModule],
            declarations: [
                NavigationMenuComponent,
                NavigationMenuItemComponent,
                NavigationMenuSubmenuComponent,
                NavigationMenuSubmenuItemComponent
            ],
            exports: [
                NavigationMenuComponent,
                NavigationMenuItemComponent,
                NavigationMenuSubmenuComponent,
                NavigationMenuSubmenuItemComponent
            ]
        })
    ], NavigationMenuModule);
    return NavigationMenuModule;
}());
export { NavigationMenuModule };
export { NavigationMenuComponent, NavigationMenuItemComponent, NavigationMenuSubmenuComponent, NavigationMenuSubmenuItemComponent };
