// @bmw-ds/components/esm5/text-link/text-link.component.js
import { __decorate } from "tslib";
import { Component, HostBinding, Input } from '@angular/core';
/* tslint:disable:component-selector*/
var TextLinkComponent = /** @class */ (function () {
    function TextLinkComponent() {
        this.bmwTextLink = true;
        this.standalone = false;
        this.id = Math.random()
            .toString(36)
            .substring(2);
    }
    __decorate([
        HostBinding('class.bmw-text-link')
    ], TextLinkComponent.prototype, "bmwTextLink", void 0);
    __decorate([
        HostBinding('class.standalone'),
        Input()
    ], TextLinkComponent.prototype, "standalone", void 0);
    __decorate([
        HostBinding('attr.id'),
        Input()
    ], TextLinkComponent.prototype, "id", void 0);
    TextLinkComponent = __decorate([
        Component({
            selector: '[bmwTextLink]',
            template: "<ng-content></ng-content>\n",
            styles: ["::ng-deep a.bmw-text-link{font-weight:700;color:var(--text-link__default__color)}::ng-deep a.bmw-text-link:focus{outline:solid 1px;outline-color:var(--color-bmw-highlight)}::ng-deep a.bmw-text-link:hover{color:var(--text-link__hover__color)}::ng-deep a.bmw-text-link.standalone{display:block;margin-bottom:var(--text-link__standalone__margin-bottom);font-family:Arial;font-size:var(--text-link__standalone__font-size);line-height:var(--text-link__standalone__line-height)}::ng-deep a.bmw-text-link.standalone::before{margin-right:var(--spacing-bmw-xs);font-family:iwp;font-weight:400;content:'\\e96a';text-decoration:none;display:inline-block}"]
        })
    ], TextLinkComponent);
    return TextLinkComponent;
}());
export { TextLinkComponent };
