// @bmw-ds/components/esm5/tooltip/tooltip.component.js
import { __decorate } from "tslib";
import { Component, HostListener, Input, ViewChild } from '@angular/core';
import { TooltipPositionService } from './tooltip-position.service';
var TooltipComponent = /** @class */ (function () {
    function TooltipComponent(tooltipPositionService) {
        this.tooltipPositionService = tooltipPositionService;
        this.position = "below" /* BELOW */;
        this.display = "inline" /* INLINE */;
        this.active = false;
        this.disabled = false;
    }
    TooltipComponent.prototype.show = function () {
        var _this = this;
        this.active = true;
        // Use setTimeout because tooltip in not present in the DOM to calculate it's position
        setTimeout(function () {
            _this.position = _this.tooltipPositionService.switchPosition(_this.tooltipRef, _this.position);
        }, 0);
    };
    TooltipComponent.prototype.hide = function () {
        this.active = false;
    };
    TooltipComponent.prototype.onMouseEnter = function () {
        this.show();
    };
    TooltipComponent.prototype.onMouseLeave = function () {
        this.hide();
    };
    TooltipComponent.ctorParameters = function () { return [
        { type: TooltipPositionService }
    ]; };
    __decorate([
        Input()
    ], TooltipComponent.prototype, "theme", void 0);
    __decorate([
        Input()
    ], TooltipComponent.prototype, "heading", void 0);
    __decorate([
        Input()
    ], TooltipComponent.prototype, "content", void 0);
    __decorate([
        Input()
    ], TooltipComponent.prototype, "position", void 0);
    __decorate([
        Input()
    ], TooltipComponent.prototype, "display", void 0);
    __decorate([
        Input()
    ], TooltipComponent.prototype, "active", void 0);
    __decorate([
        Input()
    ], TooltipComponent.prototype, "disabled", void 0);
    __decorate([
        ViewChild('tooltip')
    ], TooltipComponent.prototype, "tooltipRef", void 0);
    __decorate([
        HostListener('focus'),
        HostListener('focusin'),
        HostListener('mouseenter')
    ], TooltipComponent.prototype, "onMouseEnter", null);
    __decorate([
        HostListener('blur'),
        HostListener('focusout'),
        HostListener('mouseleave')
    ], TooltipComponent.prototype, "onMouseLeave", null);
    TooltipComponent = __decorate([
        Component({
            selector: 'bmw-tooltip',
            template: "<span class=\"bmw-tooltip\" tabindex=\"-1\" [ngClass]=\"display\">\n  <span class=\"bmw-tooltip__content\">\n    <ng-content></ng-content>\n  </span>\n  <span\n    #tooltip\n    [class.visible]=\"active && !disabled\"\n    [class.bmw-tooltip--theme-error]=\"theme === 'error'\"\n    [ngClass]=\"'bmw-tooltip-overlay--' + position\"\n    class=\"bmw-tooltip__wrapper\"\n  >\n    <ng-container *ngIf=\"heading\">\n      <div class=\"bmw-tooltip__wrapper-headline bmw-component-headline-text-small\">{{ heading }}</div>\n    </ng-container>\n    <ng-container *ngIf=\"content\">\n      <div class=\"bmw-tooltip__wrapper-content bmw-component-text\">{{ content }}</div>\n    </ng-container>\n  </span>\n</span>\n",
            styles: [".bmw-tooltip{outline:0;position:relative}.bmw-tooltip.inline{display:inline-block}.bmw-tooltip.block{display:block}.bmw-tooltip__wrapper{visibility:hidden;text-align:left;text-transform:none;background:var(--tooltip__host--background);padding:var(--tooltip__host--padding);pointer-events:auto;box-sizing:border-box;width:-webkit-max-content;width:-moz-max-content;width:max-content;min-width:var(--tooltip__host--min-width);max-width:var(--tooltip__host--max-width);max-height:var(--tooltip__host--max-height);position:absolute;z-index:var(--tooltip__z-index-base);opacity:0}.bmw-tooltip__wrapper-content{font-weight:400;color:var(--tooltip__host--color);width:100%;text-overflow:ellipsis;overflow:hidden}.bmw-tooltip__wrapper-headline{color:var(--tooltip__host--color);width:100%;text-overflow:ellipsis;overflow:hidden}.bmw-tooltip__wrapper.visible{visibility:visible;opacity:1;transition:.2s}.bmw-tooltip__wrapper.bmw-tooltip--theme-error{background:var(--tooltip__error--background)}.bmw-tooltip__wrapper:before{content:' ';border:solid transparent;height:0;width:0;position:absolute;border-width:var(--tooltip__arrow--size)}.bmw-tooltip__wrapper.bmw-tooltip-overlay--above{bottom:calc(100% + (var(--tooltip__space-offset)/ 1.5));left:50%;transform:translateX(-50%)}.bmw-tooltip__wrapper.bmw-tooltip-overlay--above:before{top:100%;left:50%;border-color:var(--tooltip__host--background) transparent transparent transparent;transform:translateX(-50%)}.bmw-tooltip__wrapper.bmw-tooltip-overlay--above.visible{bottom:calc(100% + var(--tooltip__space-offset))}.bmw-tooltip__wrapper.bmw-tooltip-overlay--above.bmw-tooltip--theme-error:before{border-top-color:var(--tooltip__error--background)}.bmw-tooltip__wrapper.bmw-tooltip-overlay--below{top:calc(100% + (var(--tooltip__space-offset)/ 1.5));left:50%;transform:translateX(-50%)}.bmw-tooltip__wrapper.bmw-tooltip-overlay--below:before{bottom:100%;left:50%;border-color:transparent transparent var(--tooltip__host--background) transparent;transform:translateX(-50%)}.bmw-tooltip__wrapper.bmw-tooltip-overlay--below.visible{top:calc(100% + var(--tooltip__space-offset))}.bmw-tooltip__wrapper.bmw-tooltip-overlay--below.bmw-tooltip--theme-error:before{border-bottom-color:var(--tooltip__error--background)}.bmw-tooltip__wrapper.bmw-tooltip-overlay--after{top:50%;transform:translateY(-50%);left:calc(100% + (var(--tooltip__space-offset)/ 1.5))}.bmw-tooltip__wrapper.bmw-tooltip-overlay--after:before{right:100%;top:50%;border-color:transparent var(--tooltip__host--background) transparent transparent;transform:translateY(-50%)}.bmw-tooltip__wrapper.bmw-tooltip-overlay--after.visible{left:calc(100% + var(--tooltip__space-offset))}.bmw-tooltip__wrapper.bmw-tooltip-overlay--after.bmw-tooltip--theme-error:before{border-right-color:var(--tooltip__error--background)}.bmw-tooltip__wrapper.bmw-tooltip-overlay--before{top:50%;transform:translateY(-50%);right:calc(100% + (var(--tooltip__space-offset)/ 1.5))}.bmw-tooltip__wrapper.bmw-tooltip-overlay--before:before{left:100%;top:50%;border-color:transparent transparent transparent var(--tooltip__host--background);transform:translateY(-50%)}.bmw-tooltip__wrapper.bmw-tooltip-overlay--before.visible{right:calc(100% + var(--tooltip__space-offset))}.bmw-tooltip__wrapper.bmw-tooltip-overlay--before.bmw-tooltip--theme-error:before{border-left-color:var(--tooltip__error--background)}"]
        })
    ], TooltipComponent);
    return TooltipComponent;
}());
export { TooltipComponent };
