// @bmw-ds/components/esm5/combo-box/combo-box-dropdown-checkbox-item/combo-box-dropdown-checkbox-item.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, HostBinding, Input, Output, ViewChild } from '@angular/core';
var ComboBoxDropdownCheckboxItemComponent = /** @class */ (function () {
    function ComboBoxDropdownCheckboxItemComponent() {
        this.disabled = false;
        this.selected = false;
        this.focused = false;
        this.id = Math.random()
            .toString(36)
            .substring(2);
        this.selectedChange = new EventEmitter();
    }
    ComboBoxDropdownCheckboxItemComponent.prototype.ngOnChanges = function () {
        var _a;
        if (this.focused) {
            (_a = this.itemWrapper) === null || _a === void 0 ? void 0 : _a.nativeElement.focus();
        }
    };
    ComboBoxDropdownCheckboxItemComponent.prototype.click = function () {
        if (!this.disabled) {
            this.selectedChange.emit(!this.selected);
        }
    };
    __decorate([
        Input()
    ], ComboBoxDropdownCheckboxItemComponent.prototype, "value", void 0);
    __decorate([
        Input()
    ], ComboBoxDropdownCheckboxItemComponent.prototype, "label", void 0);
    __decorate([
        Input()
    ], ComboBoxDropdownCheckboxItemComponent.prototype, "disabled", void 0);
    __decorate([
        Input()
    ], ComboBoxDropdownCheckboxItemComponent.prototype, "selected", void 0);
    __decorate([
        Input()
    ], ComboBoxDropdownCheckboxItemComponent.prototype, "focused", void 0);
    __decorate([
        ViewChild('itemWrapper')
    ], ComboBoxDropdownCheckboxItemComponent.prototype, "itemWrapper", void 0);
    __decorate([
        HostBinding('attr.id'),
        Input()
    ], ComboBoxDropdownCheckboxItemComponent.prototype, "id", void 0);
    __decorate([
        Output()
    ], ComboBoxDropdownCheckboxItemComponent.prototype, "selectedChange", void 0);
    ComboBoxDropdownCheckboxItemComponent = __decorate([
        Component({
            selector: 'bmw-combo-box-dropdown-checkbox-item',
            template: "<div\n  #itemWrapper\n  [class.disabled]=\"disabled\"\n  class=\"wrapper\"\n  [attr.aria-label]=\"'Option with label ' + label\"\n  tabindex=\"-1\"\n>\n  <div [class.selected]=\"selected\" [class.focused]=\"focused\" (click)=\"click()\" class=\"dropdown-item\">\n    <div class=\"label bmw-component-text\">\n      <bmw-checkbox [ngModel]=\"selected\">\n        {{ label }}\n      </bmw-checkbox>\n    </div>\n  </div>\n</div>\n",
            styles: ["div.dropdown-item{width:100%;overflow:hidden;padding:var(--combo-box__default__padding);box-sizing:border-box;position:relative;background-color:var(--combo-box__dropdown-item__default__background-color)}div.dropdown-item div.label{width:100%;display:block;box-sizing:border-box;position:relative;text-overflow:ellipsis;overflow:hidden;white-space:nowrap}div.dropdown-item div.label.bmw-component-text{color:inherit}div.dropdown-item.focused,div.dropdown-item:hover{background-color:var(--combo-box__dropdown-item__hover__background-color)}div.dropdown-item.selected{background-color:var(--combo-box__dropdown-item__selected__background-color)}div.dropdown-item.selected.focused{outline:solid 1px;outline-color:var(--color-bmw-basic5);outline-offset:-3px}div.dropdown-item.selected ::ng-deep bmw-checkbox label .bmw-component-text{color:var(--combo-box__dropdown-item__selected__color)}.disabled div.dropdown-item{pointer-events:none}.disabled div.dropdown-item .bmw-component-text{opacity:.5}div.wrapper.keyboard-focus .dropdown-item{background-color:var(--combo-box__dropdown-item__hover__background-color)}div.wrapper.disabled{cursor:no-drop}:host ::ng-deep bmw-checkbox div{margin-bottom:0}"]
        })
    ], ComboBoxDropdownCheckboxItemComponent);
    return ComboBoxDropdownCheckboxItemComponent;
}());
export { ComboBoxDropdownCheckboxItemComponent };
