// @bmw-ds/components/esm2015/master-detail-list/master-detail-list-row/master-detail-list-row.component.js
import { __decorate } from "tslib";
import { Component, ContentChildren, ElementRef, EventEmitter, HostBinding, HostListener, Input, OnInit, Output, QueryList } from '@angular/core';
import { MasterDetailListRowEntryComponent } from '../master-detail-list-row-entry/master-detail-list-row-entry.component';
let MasterDetailListRowComponent = class MasterDetailListRowComponent {
    constructor(element) {
        this.element = element;
        this.updateEvent = new EventEmitter();
    }
    set selected(value) {
        if (this._selected !== value) {
            this._selected = value;
            if (this._selected) {
                this.updateScrollPosition();
            }
        }
    }
    get selected() {
        return this._selected;
    }
    click(event) {
        this.updateEvent.emit();
    }
    ngOnInit() {
        if (!this.id)
            this.id = Math.random()
                .toString(36)
                .substring(2);
    }
    updateScrollPosition() {
        this.element.nativeElement.scrollIntoView({ behavior: 'smooth', block: 'nearest', inline: 'nearest' });
    }
};
MasterDetailListRowComponent.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Input(), HostBinding('class.selected')
], MasterDetailListRowComponent.prototype, "selected", null);
__decorate([
    ContentChildren(MasterDetailListRowEntryComponent)
], MasterDetailListRowComponent.prototype, "listEntities", void 0);
__decorate([
    HostBinding('attr.id'),
    Input()
], MasterDetailListRowComponent.prototype, "id", void 0);
__decorate([
    Output()
], MasterDetailListRowComponent.prototype, "updateEvent", void 0);
__decorate([
    HostListener('click', ['$event'])
], MasterDetailListRowComponent.prototype, "click", null);
MasterDetailListRowComponent = __decorate([
    Component({
        selector: 'bmw-master-detail-list-row',
        template: "<ng-content select=\"bmw-master-detail-list-row-entry\"></ng-content>\n",
        styles: [":host{background-color:var(--list-item__default__background-color);display:table-row-group;display:-ms-grid;display:grid;-ms-grid-columns:var(--list__item__col-size);grid-template-columns:var(--list__item__col-size);border-bottom:1px solid var(--list__item__border-bottom-color);padding:var(--spacing-bmw-xxs) var(--spacing-bmw-xs) var(--spacing-bmw-xs) var(--spacing-bmw-xs);cursor:pointer}:host:hover{background-color:var(--list__item__hover__background-color)}:host.selected{background-color:var(--list__item__selected__background-color)}@media screen and (min-width:37.49rem){:host{padding:0;display:table-row}}:host-context(bmw-master-detail-list.master-detail){display:-ms-grid;display:grid;-ms-grid-columns:var(--list__item__col-size);grid-template-columns:var(--list__item__col-size);border-bottom:1px solid var(--list__item__border-bottom-color);padding:var(--spacing-bmw-xxs) var(--spacing-bmw-xs) var(--spacing-bmw-xs) var(--spacing-bmw-xs)}"]
    })
], MasterDetailListRowComponent);
export { MasterDetailListRowComponent };
