// @bmw-ds/components/esm2015/fieldset/fieldset.component.js
import { __decorate } from "tslib";
import { Component, ContentChild, HostBinding, Input, HostListener, ViewChild } from '@angular/core';
import { FieldsetHeaderComponent } from './fieldset-header.component';
let FieldsetComponent = class FieldsetComponent {
    constructor() {
        this._disabled = false;
        this._opened = false;
    }
    get disabled() {
        return this._disabled;
    }
    set disabled(value) {
        this._disabled = value;
        if (!this.header)
            return;
        this.header.disabled = value;
    }
    get opened() {
        return this._opened;
    }
    set opened(value) {
        this._opened = value;
        if (!this.header)
            return;
        this.header.opened = value;
    }
    onKeydown(event) {
        const key = event.code || event.keyCode;
        switch (key) {
            case 'Enter':
            case 'NumpadEnter':
            case 13:
            case 'Space':
            case 32:
                if (!this.disabled && event.target === this.label.nativeElement) {
                    this.toggle();
                    event.preventDefault();
                    event.stopPropagation();
                }
                break;
        }
    }
    ngOnInit() {
        if (this.header) {
            this.header.opened = this.opened;
        }
        if (!this.id) {
            this.id = Math.random()
                .toString(36)
                .substring(2);
        }
    }
    ngAfterViewInit() {
        if (this.header) {
            this.header.clickEmit.emit = () => {
                this.toggle();
            };
        }
    }
    ngAfterContentInit() {
        if (this.header) {
            this.header.opened = this.opened;
            this.header.disabled = this.disabled;
        }
    }
    toggle() {
        if (this.disabled)
            return;
        this.opened = !this.opened;
        if (!this.header)
            return;
        this.header.opened = this.opened;
    }
};
__decorate([
    ContentChild(FieldsetHeaderComponent)
], FieldsetComponent.prototype, "header", void 0);
__decorate([
    ViewChild('label')
], FieldsetComponent.prototype, "label", void 0);
__decorate([
    Input()
], FieldsetComponent.prototype, "disabled", null);
__decorate([
    HostBinding('attr.id'),
    Input()
], FieldsetComponent.prototype, "id", void 0);
__decorate([
    Input()
], FieldsetComponent.prototype, "opened", null);
__decorate([
    HostListener('keydown', ['$event'])
], FieldsetComponent.prototype, "onKeydown", null);
FieldsetComponent = __decorate([
    Component({
        selector: 'bmw-fieldset',
        template: "<label #label [class.disabled]=\"disabled\" [tabindex]=\"disabled ? -1 : 0\"\n  ><ng-content select=\"bmw-fieldset-header\"></ng-content\n></label>\n<div [class.disabled]=\"disabled\" class=\"fieldset-container\" *ngIf=\"opened\">\n  <ng-content select=\"bmw-fieldset-container\"></ng-content>\n</div>\n",
        styles: [".fieldset-container{margin-top:var(--fieldset__container__margin-top)}label{display:block}label.disabled{pointer-events:none;opacity:.5}label:focus{outline:solid 1px;outline-color:var(--color-bmw-highlight);outline-offset:-1px}:host[disabled]{cursor:no-drop}"]
    })
], FieldsetComponent);
export { FieldsetComponent };
