// @bmw-ds/components/esm2015/fieldset/fieldset-header.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Output } from '@angular/core';
let FieldsetHeaderComponent = class FieldsetHeaderComponent {
    constructor() {
        this.opened = false;
        this.clickEmit = new EventEmitter();
    }
    emitClick() {
        this.clickEmit.emit();
    }
};
__decorate([
    Output()
], FieldsetHeaderComponent.prototype, "clickEmit", void 0);
FieldsetHeaderComponent = __decorate([
    Component({
        selector: 'bmw-fieldset-header',
        template: "<div class=\"fieldset-header\" (click)=\"emitClick()\" [class.opened]=\"opened\">\n  <div class=\"header bmw-component-headline-text\" [attr.aria-disabled]=\"disabled\"><ng-content></ng-content></div>\n  <div class=\"arrow\"></div>\n</div>\n",
        styles: [".fieldset-header{display:inline-flex;max-width:100%;cursor:pointer}.fieldset-header .arrow{order:2}.fieldset-header .arrow::after{padding:calc((var(--typography__component-headline-text__line-height) - var(--typography__component-headline-text__font-size))/ 2);padding-left:var(--fieldset__header__arrow__padding-left);height:var(--typography__component-headline-text__line-height);width:var(--typography__component-headline-text__line-height);font-size:var(--typography__component-headline-text__font-size);font-family:iwp;display:block;content:'\\e92f'}.fieldset-header .header{text-overflow:ellipsis;white-space:nowrap;overflow:hidden}.fieldset-header.opened .arrow::after{content:'\\e932'}.fieldset-header:hover .bmw-component-headline-text{color:var(--fieldset__header__hover__color)}"]
    })
], FieldsetHeaderComponent);
export { FieldsetHeaderComponent };
