// @bmw-ds/components/esm2015/bottom-navigation/bottom-navigation-item.component.js
import { __decorate } from "tslib";
import { Component, ElementRef, HostBinding, HostListener, Input, OnInit, Output } from '@angular/core';
import { Subject } from 'rxjs';
let BottomNavigationItemComponent = class BottomNavigationItemComponent {
    constructor(el) {
        this.el = el;
        this._selected = false;
        this.tabIndex = 0;
        this.selectedChange = new Subject();
    }
    set selected(value) {
        if (value === this._selected)
            return;
        this._selected = value;
        if (this._selected) {
            this.selectedChange.next(this._selected);
        }
    }
    get selected() {
        return this._selected;
    }
    ngOnInit() {
        if (!this.id) {
            this.id = Math.random()
                .toString(36)
                .substring(2);
        }
    }
    click() {
        this._selected = !this._selected;
        this.selectedChange.next(this._selected);
    }
    get nativeElement() {
        return this.el.nativeElement;
    }
};
BottomNavigationItemComponent.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Input()
], BottomNavigationItemComponent.prototype, "icon", void 0);
__decorate([
    HostBinding('title'),
    Input()
], BottomNavigationItemComponent.prototype, "label", void 0);
__decorate([
    HostBinding('class.selected'),
    Input()
], BottomNavigationItemComponent.prototype, "selected", null);
__decorate([
    HostBinding('attr.id'),
    Input()
], BottomNavigationItemComponent.prototype, "id", void 0);
__decorate([
    HostBinding('tabindex')
], BottomNavigationItemComponent.prototype, "tabIndex", void 0);
__decorate([
    Output()
], BottomNavigationItemComponent.prototype, "selectedChange", void 0);
__decorate([
    HostListener('click')
], BottomNavigationItemComponent.prototype, "click", null);
BottomNavigationItemComponent = __decorate([
    Component({
        selector: 'bmw-bottom-navigation-item',
        template: "<i [ngClass]=\"icon\"> </i>\n<span class=\"label bmw-small-text\" [textContent]=\"label\"></span>\n",
        styles: [":host{display:flex;flex-direction:column;align-items:center;justify-content:center;min-height:var(--bottom-navigation-item__min-height);min-width:calc(var(--bottom-navigation-item__min-width) - (2 * var(--bottom-navigation-item__padding)));padding:0 var(--bottom-navigation-item__padding);cursor:pointer}:host:focus{outline:solid 1px;outline-color:var(--color-bmw-highlight);outline-offset:-1px}:host:focus.selected:not(.secondary){outline:solid 1px;outline-color:var(--color-bmw-basic5);outline-offset:-3px}:host:hover:not(.selected){background-color:var(--bottom-navigation__hover__background-color)}:host.selected:not(.secondary){background-color:var(--bottom-navigation__selected__background-color)}:host.selected:not(.secondary) i,:host.selected:not(.secondary) span{color:var(--bottom-navigation__selected__color)}:host.secondary.child-selected:not(.selected){background-color:var(--bottom-navigation__selected__background-color)}:host.secondary.child-selected:not(.selected):focus{outline:solid 1px;outline-color:var(--color-bmw-basic5);outline-offset:-3px}:host.secondary.child-selected:not(.selected) i,:host.secondary.child-selected:not(.selected) span{color:var(--bottom-navigation__selected__color)}:host.secondary.selected{background-color:var(--bottom-navigation__secondary-selected__color)}:host i{color:var(--bottom-navigation__default__color);font-family:iwp;font-size:var(--bottom-navigation-item__icon__font-size);height:var(--bottom-navigation-item__icon__font-size);text-align:center;width:var(--bottom-navigation-item__icon__font-size)}:host span{margin-top:var(--bottom-navigation-item__free-space);white-space:nowrap;text-overflow:ellipsis;max-width:100%;overflow:hidden}:host-context(bmw-bottom-navigation.base-container){flex:1 1 100%}"]
    })
], BottomNavigationItemComponent);
export { BottomNavigationItemComponent };
