// @bmw-ds/components/esm5/split-button/split-button.component.js
import { __decorate } from "tslib";
import { Component, Input, ViewContainerRef, ViewChild, ContentChild, HostBinding, AfterViewInit, EventEmitter, Output, HostListener, ElementRef } from '@angular/core';
import { ContextMenuComponent } from '../context-menu/context-menu.component';
var SplitButtonComponent = /** @class */ (function () {
    function SplitButtonComponent(elemRef) {
        this.elemRef = elemRef;
        this.itemSelected = new EventEmitter();
        this.clickEvent = new EventEmitter();
    }
    SplitButtonComponent.prototype.onKeydown = function (event) {
        var key = event.code || event.keyCode;
        switch (key) {
            case 'Enter':
            case 'NumpadEnter':
            case 27:
            case 'Space':
            case 32:
                if (event.target === this.elemRef.nativeElement.querySelector('.button-arrow') ||
                    event.target === this.elemRef.nativeElement.querySelector('.icon-arrow')) {
                    this.toggle();
                    event.preventDefault();
                    event.stopPropagation();
                }
                break;
        }
    };
    SplitButtonComponent.prototype.clickout = function (event) {
        if (!this.contextMenu.element.nativeElement.contains(event.target)) {
            this.toggle(false);
        }
        event.stopPropagation();
    };
    SplitButtonComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        this.childContextMenu.childContextMenuItems.forEach(function (item) {
            _this.assignClick(item);
        });
    };
    // Recursevly shake all menu items and emit item onClick and close menu
    SplitButtonComponent.prototype.assignClick = function (menuItem) {
        var _this = this;
        if (menuItem.subMenu) {
            menuItem.subMenu.childContextMenuItems.forEach(function (item) {
                _this.assignClick(item);
            });
            return;
        }
        menuItem.selected.subscribe(function (item) {
            _this.itemSelected.emit(item);
            _this.toggle(false);
        });
    };
    SplitButtonComponent.prototype.arrowButtonClick = function (event) {
        this.toggle();
        event === null || event === void 0 ? void 0 : event.stopPropagation();
    };
    SplitButtonComponent.prototype.buttonClick = function () {
        if (this.disabledButton)
            return;
        this.clickEvent.emit(this);
    };
    SplitButtonComponent.prototype.toggle = function (state) {
        if (!this.childContextMenu || !this.childContextMenu.toggle)
            return;
        this.childContextMenu.unselectSubs();
        this.childContextMenu.toggle(state);
    };
    Object.defineProperty(SplitButtonComponent.prototype, "focus", {
        get: function () {
            return this.childContextMenu ? this.childContextMenu.opened : false;
        },
        enumerable: true,
        configurable: true
    });
    SplitButtonComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        ContentChild(ContextMenuComponent)
    ], SplitButtonComponent.prototype, "childContextMenu", void 0);
    __decorate([
        ViewChild('contextMenu', { read: ViewContainerRef })
    ], SplitButtonComponent.prototype, "contextMenu", void 0);
    __decorate([
        Input()
    ], SplitButtonComponent.prototype, "iconClass", void 0);
    __decorate([
        Input()
    ], SplitButtonComponent.prototype, "disabledButton", void 0);
    __decorate([
        Input()
    ], SplitButtonComponent.prototype, "disabledMenu", void 0);
    __decorate([
        Output()
    ], SplitButtonComponent.prototype, "itemSelected", void 0);
    __decorate([
        Output()
    ], SplitButtonComponent.prototype, "clickEvent", void 0);
    __decorate([
        HostListener('keydown', ['$event'])
    ], SplitButtonComponent.prototype, "onKeydown", null);
    __decorate([
        HostListener('document:click', ['$event'])
    ], SplitButtonComponent.prototype, "clickout", null);
    __decorate([
        HostBinding('class.opened')
    ], SplitButtonComponent.prototype, "focus", null);
    SplitButtonComponent = __decorate([
        Component({
            selector: 'bmw-split-button',
            template: "<div #contextMenu>\n  <div class=\"icon-wrapper\" *ngIf=\"iconClass\">\n    <bmw-icon-button\n      [iconClass]=\"iconClass\"\n      (clickEvent)=\"buttonClick()\"\n      [disabled]=\"disabledButton\"\n      [tabIndex]=\"disabledButton ? -1 : 0\"\n    ></bmw-icon-button>\n    <div class=\"icon-arrow\" (click)=\"toggle()\" [class.disabled]=\"disabledMenu\" [tabindex]=\"disabledMenu ? -1 : 0\"></div>\n  </div>\n  <div class=\"button-wrapper\" *ngIf=\"!iconClass\">\n    <bmw-button\n      (click)=\"buttonClick()\"\n      [attr.disabled]=\"disabledButton\"\n      [disabled]=\"disabledButton\"\n      [tabIndex]=\"disabledButton ? -1 : 0\"\n    >\n      <ng-content select=\"label\"></ng-content>\n    </bmw-button>\n    <div\n      class=\"button-arrow\"\n      (click)=\"arrowButtonClick($event)\"\n      [class.disabled]=\"disabledMenu\"\n      [tabindex]=\"disabledMenu ? -1 : 0\"\n    ></div>\n  </div>\n  <ng-content select=\"bmw-context-menu\"></ng-content>\n</div>\n",
            styles: ["div:focus{outline:0}div.button-wrapper,div.icon-wrapper{position:relative;display:inline-block;vertical-align:middle;max-width:100%}div.button-wrapper div.button-arrow,div.button-wrapper div.icon-arrow,div.icon-wrapper div.button-arrow,div.icon-wrapper div.icon-arrow{cursor:pointer}div.button-wrapper div.button-arrow.disabled,div.button-wrapper div.icon-arrow.disabled,div.icon-wrapper div.button-arrow.disabled,div.icon-wrapper div.icon-arrow.disabled{cursor:no-drop}div.button-wrapper div.button-arrow.disabled:active::after,div.button-wrapper div.icon-arrow.disabled:active::after,div.icon-wrapper div.button-arrow.disabled:active::after,div.icon-wrapper div.icon-arrow.disabled:active::after{pointer-events:none}div.button-wrapper div.button-arrow.disabled::after,div.button-wrapper div.icon-arrow.disabled::after,div.icon-wrapper div.button-arrow.disabled::after,div.icon-wrapper div.icon-arrow.disabled::after{opacity:.7}div.button-wrapper div.button-arrow::after,div.button-wrapper div.icon-arrow::after,div.icon-wrapper div.button-arrow::after,div.icon-wrapper div.icon-arrow::after{content:'\\e92f';font-family:iwp;position:absolute;right:0;color:var(--split-button__default__color)}div.button-wrapper div.button-arrow:hover::after,div.button-wrapper div.icon-arrow:hover::after,div.icon-wrapper div.button-arrow:hover::after,div.icon-wrapper div.icon-arrow:hover::after{background-color:var(--split-button__selected__background-color)}div.button-wrapper div.button-arrow:focus::after,div.button-wrapper div.icon-arrow:focus::after,div.icon-wrapper div.button-arrow:focus::after,div.icon-wrapper div.icon-arrow:focus::after{outline:solid 1px;outline-color:var(--color-bmw-highlight);outline-offset:-1px}div.icon-wrapper{padding-right:var(--split-button__icon__padding-right)}div.icon-wrapper div.icon-arrow::after{width:var(--split-button__icon__arrow__font-size);height:var(--split-button__icon__arrow__font-size);padding:var(--split-button__icon-inner__padding-vertical) var(--split-button__icon-inner__padding-horizontal);font-size:var(--split-button__icon__arrow__font-size);top:0}@media all and (-ms-high-contrast:none),(-ms-high-contrast:active){div.icon-wrapper div.icon-arrow::after{padding:10px 3px}}div.icon-wrapper div.icon-arrow.disabled:hover::after{background-color:initial}div.button-wrapper{padding-right:var(--split-button__button__padding-right)}div.button-wrapper div.button-arrow::after{width:var(--split-button__icon__arrow__width);height:var(--split-button__icon__arrow__height);padding:var(--split-button__button_inner__padding);font-size:var(--split-button__button__arrow__font-size);line-height:var(--split-button__icon__arrow__height);text-align:center;top:0;background-color:var(--split-button__default__background-color);border:solid 1px var(--split-button__default__border-color)}div.button-wrapper div.button-arrow.disabled:hover::after{background-color:var(--split-button__default__background-color)}:host{max-width:calc(100% - var(--split-button__button__padding-right));display:inline-block;position:relative}:host ::ng-deep label{cursor:inherit;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}:host ::ng-deep .bmw-context-menu{min-width:100%}:host.opened div.button-wrapper div.button-arrow::after,:host.opened div.button-wrapper div.icon-arrow::after,:host.opened div.icon-wrapper div.button-arrow::after,:host.opened div.icon-wrapper div.icon-arrow::after{content:'\\e932';background-color:var(--split-button__selected__background-color)}"]
        })
    ], SplitButtonComponent);
    return SplitButtonComponent;
}());
export { SplitButtonComponent };
