// @bmw-ds/components/esm5/button/button.component.js
import { __decorate } from "tslib";
import { Component, Input, OnInit, Output, EventEmitter, ElementRef, OnChanges } from '@angular/core';
var ButtonComponent = /** @class */ (function () {
    function ButtonComponent(elementRef) {
        this.elementRef = elementRef;
        this.clickEvent = new EventEmitter();
        this.primary = false;
        this.secondary = false;
        this.small = false;
        this.supportedTypes = ['submit', 'reset', 'button'];
    }
    ButtonComponent.prototype.ngOnInit = function () {
        if (!this.id)
            this.id = Math.random()
                .toString(36)
                .substring(2);
        this.checkNxProperties();
    };
    ButtonComponent.prototype.ngOnChanges = function () {
        this.checkNxProperties();
    };
    ButtonComponent.prototype.checkNxProperties = function () {
        if (this.button) {
            var properties = this.button.split(' ');
            if (properties.indexOf('primary') > -1) {
                this.primary = true;
            }
            else if (properties.indexOf('secondary') > -1) {
                this.secondary = true;
            }
            if (properties.indexOf('small') > -1) {
                this.small = true;
            }
        }
        if (this.type && !this.supportedTypes.includes(this.type)) {
            console.warn('Wrong button type!', this);
            this.type = 'button';
        }
        if (this.iconLeft) {
            this.iconRight = undefined;
        }
        if (this.disabled) {
            this.elementRef.nativeElement.classList.add('disabled');
        }
        else {
            this.elementRef.nativeElement.classList.remove('disabled');
        }
    };
    ButtonComponent.prototype.click = function () {
        if (this.disabled)
            return;
        this.clickEvent.emit(this);
    };
    ButtonComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Input()
    ], ButtonComponent.prototype, "button", void 0);
    __decorate([
        Input()
    ], ButtonComponent.prototype, "iconRight", void 0);
    __decorate([
        Input()
    ], ButtonComponent.prototype, "iconLeft", void 0);
    __decorate([
        Input()
    ], ButtonComponent.prototype, "type", void 0);
    __decorate([
        Input()
    ], ButtonComponent.prototype, "disabled", void 0);
    __decorate([
        Input()
    ], ButtonComponent.prototype, "id", void 0);
    __decorate([
        Input()
    ], ButtonComponent.prototype, "tabIndex", void 0);
    __decorate([
        Output()
    ], ButtonComponent.prototype, "clickEvent", void 0);
    ButtonComponent = __decorate([
        Component({
            selector: 'bmw-button',
            template: "<button\n  class=\"button\"\n  [class.icon-left]=\"iconLeft\"\n  [class.icon-right]=\"iconRight\"\n  [class.button--primary]=\"primary === true\"\n  [class.button--secondary]=\"secondary === true\"\n  [class.button--small]=\"small === true\"\n  [disabled]=\"disabled\"\n  [type]=\"type\"\n  (click)=\"click()\"\n  [id]=\"id\"\n  [tabindex]=\"disabled ? -1 : tabIndex\"\n>\n  <i *ngIf=\"iconLeft\" [ngClass]=\"[iconLeft, 'icon-left']\"></i>\n  <div [ngClass]=\"small === true ? 'bmw-component-headline-text-small' : 'bmw-component-headline-text'\">\n    <ng-content></ng-content>\n  </div>\n  <i *ngIf=\"iconRight\" [ngClass]=\"[iconRight, 'icon-right']\"></i>\n</button>\n",
            styles: [".button{padding:var(--button__medium__padding);text-transform:uppercase;color:var(--button__default__color);border:solid 1px var(--button__default__border__color);background-color:var(--button__default__background-color);position:relative;max-width:100%;cursor:pointer}.button.icon-left{padding-left:calc(var(--typography__component-headline-text__line-height) + var(--button__medium__padding) + var(--button__default__icon__padding))}.button.icon-right{padding-right:calc(var(--typography__component-headline-text__line-height) + var(--button__medium__padding) + var(--button__default__icon__padding))}@media all and (-ms-high-contrast:none),(-ms-high-contrast:active){.button{padding-top:calc(var(--button__medium__padding) + 2px);padding-bottom:calc(var(--button__medium__padding) - 2px)}.button.icon-left{padding-left:30px}.button.icon-right{padding-right:30px}}.button i{font-size:var(--typography__component-headline-text__font-size);line-height:var(--typography__component-headline-text__line-height);position:absolute;top:calc(50% - var(--typography__component-headline-text__line-height)/ 2)}.button i.icon-left{left:var(--button__medium__padding)}.button i.icon-right{right:var(--button__medium__padding)}.button:focus{outline:solid 1px;outline-color:var(--color-bmw-highlight);outline-offset:-1px;z-index:var(--z-index__content-base)}.button .bmw-component-headline-text{white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.button.button--primary{border:solid 1px var(--button__primary__background-color);background-color:var(--button__primary__background-color);color:var(--button__primary__color)}.button.button--primary .bmw-component-headline-text,.button.button--primary .bmw-component-headline-text-small{color:var(--button__primary__color)}.button.button--primary:focus{outline:solid 1px;outline-color:var(--color-bmw-basic5);outline-offset:-3px}.button.button--secondary{border:solid 1px var(--button__secondary__border__color);background-color:var(--button__secondary__background-color)}.button:active{background-color:var(--button__default__active__background-color);border:solid 1px var(--button__default__active__background-color);color:var(--button__default__active__color)}.button:active .bmw-component-headline-text,.button:active .bmw-component-headline-text-small{color:var(--button__default__active__color)}.button:active.button--primary{background-color:var(--button__primary__active__background-color)}.button:active.button--secondary{background-color:var(--button__secondary__active__background-color)}.button:active,.button:hover{background-color:var(--button__default__hover__background-color);border:solid 1px var(--button__default__hover__border-color);color:var(--button__default__hover__color)}.button:active .bmw-component-headline-text,.button:active .bmw-component-headline-text-small,.button:hover .bmw-component-headline-text,.button:hover .bmw-component-headline-text-small{color:var(--button__default__hover__color)}.button:active.button--primary,.button:hover.button--primary{color:var(--button__primary__color);background-color:var(--button__primary__hover__background-color);border:solid 1px var(--button__primary__hover__background-color)}.button:active.button--primary .bmw-component-headline-text,.button:hover.button--primary .bmw-component-headline-text{color:var(--button__primary__hover__color)}.button:active.button--primary .bmw-component-headline-text-small,.button:hover.button--primary .bmw-component-headline-text-small{color:var(--button__primary__hover__color)}.button:active.button--secondary,.button:hover.button--secondary{color:var(--button__secondary__hover__color);background-color:var(--button__secondary__hover__background-color);border:solid 1px var(--button__secondary__border__color)}.button:active.button--secondary .bmw-component-headline-text,.button:hover.button--secondary .bmw-component-headline-text{color:var(--button__secondary__hover__color)}.button:active.button--secondary .bmw-component-headline-text-small,.button:hover.button--secondary .bmw-component-headline-text-small{color:var(--button__secondary__hover__color)}.button[disabled]{opacity:.7;pointer-events:none}.button.button--small{padding:var(--button__small__padding);padding-top:var(--button__small__padding-top);padding-bottom:var(--button__small__padding-bottom)}@media all and (-ms-high-contrast:none),(-ms-high-contrast:active){.button.button--small{padding-top:calc(var(--button__small__padding-top) + 2px);padding-bottom:calc(var(--button__small__padding-bottom) - 2px)}}.button.button--small i{font-size:var(--typography__component-headline-text-small__font-size);line-height:var(--typography__component-headline-text-small__line-height);top:calc(50% - (var(--typography__component-headline-text-small__line-height)/ 2))}.button.button--small i.icon-left{left:var(--button__small__padding)}.button.button--small i.icon-right{right:var(--button__small__padding)}.button.button--small.icon-left{padding-left:calc(var(--typography__component-headline-text-small__font-size) + var(--button__small__padding) + var(--button__default__icon__padding))}.button.button--small.icon-right{padding-right:calc(var(--typography__component-headline-text-small__font-size) + var(--button__small__padding) + var(--button__default__icon__padding))}:host{max-width:100%}:host.disabled{cursor:no-drop;display:inline-block}"]
        })
    ], ButtonComponent);
    return ButtonComponent;
}());
export { ButtonComponent };
