// @bmw-ds/components/esm5/content-tab/content-tab-label.component.js
import { __decorate } from "tslib";
import { Component, ElementRef, Input, ViewChild, HostBinding, SimpleChanges, Output, EventEmitter, OnChanges } from '@angular/core';
var ContentTabLabelComponent = /** @class */ (function () {
    function ContentTabLabelComponent(elementRef) {
        this.elementRef = elementRef;
        this.change = new EventEmitter();
        // Boolean flag to pre-select this label and the respective content.
        // If multiple tabs per tabset have the selected input set to true, the last one of them will be selected.
        this.selected = false;
        this.disabled = false;
        this.role = 'tab';
    }
    ContentTabLabelComponent.prototype.ngOnChanges = function () {
        this.change.emit();
    };
    ContentTabLabelComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Input()
    ], ContentTabLabelComponent.prototype, "icon", void 0);
    __decorate([
        Input()
    ], ContentTabLabelComponent.prototype, "for", void 0);
    __decorate([
        Output()
    ], ContentTabLabelComponent.prototype, "change", void 0);
    __decorate([
        HostBinding('attr.aria-expanded'),
        HostBinding('attr.aria-selected'),
        Input()
    ], ContentTabLabelComponent.prototype, "selected", void 0);
    __decorate([
        HostBinding('attr.aria-disabled'),
        Input()
    ], ContentTabLabelComponent.prototype, "disabled", void 0);
    __decorate([
        ViewChild('label')
    ], ContentTabLabelComponent.prototype, "label", void 0);
    __decorate([
        HostBinding('attr.role')
    ], ContentTabLabelComponent.prototype, "role", void 0);
    ContentTabLabelComponent = __decorate([
        Component({
            selector: 'bmw-content-tab-label',
            template: "<label #label class=\"bmw-component-headline-text\" [tabindex]=\"selected && !disabled ? 0 : -1\">\n  <i [ngClass]=\"[icon || '']\"></i\n  ><span>\n    <ng-content></ng-content>\n  </span>\n</label>\n",
            styles: ["label{display:block;min-width:var(--content-tabs__label__default__min-width);padding:var(--content-tabs__label__default__padding);padding-bottom:calc(var(--content-tabs__label__default__padding));text-align:center;box-sizing:border-box;border-bottom:2px solid transparent;height:calc(var(--content-tabs__label__default__height) + 2px);z-index:2;cursor:pointer}label span{text-overflow:ellipsis;white-space:nowrap;overflow:hidden;display:block}label:focus{outline:solid 1px;outline-color:var(--color-bmw-highlight);outline-offset:-1px}:host.just-icon label{font-family:iwp;font-size:var(--content-tabs__label__icon__font-size);min-width:auto;padding:var(--content-tabs__label__icon__padding);padding-bottom:calc(var(--content-tabs__label__icon__padding));margin-right:var(--content-tabs__label__icon__margin-right)}:host.just-icon label i{display:flex;height:var(--content-tabs__label__icon__font-size)}:host.just-icon label span{display:none}:host.with-icon label{padding-left:var(--content-tabs__label__with-icon__padding-left);position:relative}@media all and (-ms-high-contrast:none),(-ms-high-contrast:active){:host.with-icon label{padding-left:30px}}:host.with-icon label i{position:absolute;display:block;left:var(--content-tabs__icon__with-icon__padding-left);top:calc(50% - var(--content-tabs__label__with-icon__icon-size)/ 2 + 1px);font-size:var(--content-tabs__label__with-icon__icon-size)}:host label i{display:none}:host.selected label{border-bottom:2px solid var(--content-tabs__labels__selected__color)}:host.selected label.bmw-component-headline-text{color:var(--content-tabs__labels__selected__color)}label:hover.bmw-component-headline-text{color:var(--content-tabs__labels__hover__color)}:host{display:inline-flex;max-width:100%}:host.disabled{opacity:.5;cursor:no-drop;pointer-events:none}"]
        })
    ], ContentTabLabelComponent);
    return ContentTabLabelComponent;
}());
export { ContentTabLabelComponent };
