// @bmw-ds/components/esm5/content-tab/content-tab-content.component.js
import { __decorate } from "tslib";
import { Component, Input, HostBinding } from '@angular/core';
var ContentTabContentComponent = /** @class */ (function () {
    function ContentTabContentComponent() {
        /**
         * expanding content toggle input (default = false)
         */
        this.expandContent = false;
    }
    ContentTabContentComponent.prototype.ngOnInit = function () { };
    __decorate([
        Input()
    ], ContentTabContentComponent.prototype, "hide", void 0);
    __decorate([
        HostBinding('class.expanding'), Input()
    ], ContentTabContentComponent.prototype, "expandContent", void 0);
    ContentTabContentComponent = __decorate([
        Component({
            selector: 'bmw-content-tab-content',
            template: "<div class=\"content-area\" [class.hidden]=\"hide\" role=\"tabpanel\">\n  <ng-content></ng-content>\n</div>\n",
            styles: [".content-area{display:none;height:100%}:host.hidden .content-area{display:none}:host.selected>.content-area{padding:var(--content-tabs__container__padding);display:block}:host{display:flex;flex-direction:column;min-height:0}:host.expanding{flex-grow:1;display:none;overflow:auto}:host.expanding.selected{display:flex}:host.selected>.content-area{display:flex;flex-direction:column;min-height:0}"]
        })
    ], ContentTabContentComponent);
    return ContentTabContentComponent;
}());
export { ContentTabContentComponent };
