// @bmw-ds/components/esm5/list/list-item.component.js
import { __decorate } from "tslib";
import { Component, ContentChild, HostBinding, Input, ViewChild, Output, EventEmitter } from '@angular/core';
import { ContextMenuComponent } from '../context-menu/context-menu.component';
var ListItemComponent = /** @class */ (function () {
    function ListItemComponent() {
        this.multiSelect = false;
        this.updateEvent = new EventEmitter();
        this.withContextMenu = false;
    }
    ListItemComponent.prototype.ngOnInit = function () {
        if (!this.id)
            this.id = Math.random()
                .toString(36)
                .substring(2);
    };
    ListItemComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        if (this.contextMenu.nativeElement.firstChild) {
            // wait for next tick to prevent expresionChangedAfterCheck error
            setTimeout(function () {
                _this.withContextMenu = true;
            });
        }
    };
    ListItemComponent.prototype.focusOut = function (e) {
        var leavingParent = !this.contextMenu.nativeElement.contains(e.relatedTarget);
        if (leavingParent) {
            this.toggle(false);
        }
    };
    ListItemComponent.prototype.openContextMenu = function (e) {
        this.toggle();
        e.stopPropagation();
    };
    ListItemComponent.prototype.toggle = function (state) {
        if (!this.childContextMenuComponent)
            return;
        this.childContextMenuComponent.toggle(state);
    };
    ListItemComponent.prototype.checkboxUpdate = function (event) {
        if ((event.checked === 'checked') === this.selected)
            return;
        this.updateEvent.emit(event.checked === 'checked');
    };
    ListItemComponent.prototype.click = function (event) {
        //ignore clicks inside checkbox
        if (event.target !== this.labelElement.nativeElement)
            return;
        this.updateEvent.emit(this.selected);
    };
    __decorate([
        ContentChild(ContextMenuComponent)
    ], ListItemComponent.prototype, "childContextMenuComponent", void 0);
    __decorate([
        ViewChild('contextMenu')
    ], ListItemComponent.prototype, "contextMenu", void 0);
    __decorate([
        ViewChild('labelElement')
    ], ListItemComponent.prototype, "labelElement", void 0);
    __decorate([
        ViewChild('checkboxElement', { static: false })
    ], ListItemComponent.prototype, "checkboxElement", void 0);
    __decorate([
        HostBinding('attr.id'),
        Input()
    ], ListItemComponent.prototype, "id", void 0);
    __decorate([
        Input()
    ], ListItemComponent.prototype, "selected", void 0);
    __decorate([
        Input()
    ], ListItemComponent.prototype, "icon", void 0);
    __decorate([
        Input()
    ], ListItemComponent.prototype, "multiSelect", void 0);
    __decorate([
        Input()
    ], ListItemComponent.prototype, "label", void 0);
    __decorate([
        HostBinding('style.order'),
        Input()
    ], ListItemComponent.prototype, "order", void 0);
    __decorate([
        Output()
    ], ListItemComponent.prototype, "updateEvent", void 0);
    ListItemComponent = __decorate([
        Component({
            selector: 'bmw-list-item',
            template: "<div class=\"wrapper\" [class.selected]=\"selected\" (click)=\"click($event)\">\n  <div\n    class=\"item bmw-component-text\"\n    [class.selected]=\"selected\"\n    [class.with-context-menu]=\"withContextMenu\"\n    [class.multi-select]=\"multiSelect\"\n    #labelElement\n  >\n    <div class=\"separator\" *ngIf=\"multiSelect\"></div>\n    <bmw-checkbox\n      *ngIf=\"multiSelect\"\n      #checkboxElement\n      [checked]=\"selected ? 'checked' : 'unchecked'\"\n      (updateEvent)=\"checkboxUpdate($event)\"\n    ></bmw-checkbox>\n    <i *ngIf=\"icon\" class=\"icon\" [ngClass]=\"icon\"></i>\n    <span *ngIf=\"label\">{{ label }}</span>\n    <div class=\"context-menu\" #contextMenu (focusout)=\"focusOut($event)\" tabindex=\"0\" (click)=\"openContextMenu($event)\">\n      <ng-content select=\"bmw-context-menu\"></ng-content>\n    </div>\n  </div>\n</div>\n",
            styles: [".wrapper .item{position:relative;padding:var(--list__item__padding-vertical) var(--list__item__padding-horizontal);padding-bottom:calc(var(--list__item__padding-vertical) - 1px);border-bottom:1px solid var(--list__item__border-bottom-color);box-sizing:border-box;white-space:wrap;color:var(--list__default__color)}.wrapper .item.with-context-menu{padding-right:calc(var(--list__item__padding-horizontal) + var(--list__item__padding-vertical) + var(--list__icon__height) * .125 + var(--typography__component-text__font-size))}.wrapper .item ::ng-deep label div{height:var(--checkbox__default__height)}.wrapper .item ::ng-deep bmw-checkbox{position:absolute;top:calc((var(--list__item__line-height) - var(--checkbox__default__height))/ 2 + var(--list__item__padding-vertical));left:var(--list__header__padding)}.wrapper .item.selected{background-color:var(--list__item__selected__background-color)}.wrapper .item.multi-select{padding-left:var(--list__header__selectable__padding-left)}.wrapper .item .context-menu{position:absolute;right:0;top:0;box-sizing:border-box;display:inline-block}.wrapper .item .context-menu ::ng-deep bmw-context-menu div.bmwContextMenu{top:100%;right:0}.wrapper .item .context-menu::after{content:'\\e923';font-family:iwp;height:var(--list__icon__height);display:block;padding:var(--list__item__padding-vertical) calc(var(--list__item__padding-vertical) + var(--list__icon__height) * .125)}.wrapper .item .context-menu:empty{display:none}.wrapper .item .context-menu:hover{background-color:var(--list__item__context-menu__opened__background-color)}.wrapper .item .context-menu:focus{outline:0}:host.opened .wrapper .item .context-menu{background-color:var(--list__item__context-menu__opened__background-color)}.wrapper .item .separator{position:absolute;top:0;left:0;height:100%;width:calc(var(--list__icon__height) + 2 * var(--list__header__padding));border-right:1px solid #a3a3a3}.wrapper .item .icon{padding-right:var(--spacing-bmw-xxs)}.wrapper:hover{background-color:var(--list__item__hover__background-color)}:host{display:block;background-color:var(--list-item__default__background-color)}"]
        })
    ], ListItemComponent);
    return ListItemComponent;
}());
export { ListItemComponent };
