// @bmw-ds/components/esm5/switch/switch.component.js
import { __decorate, __param } from "tslib";
import { Component, AfterViewInit, Input, ViewChild, ElementRef, Self, Optional, HostBinding } from '@angular/core';
import { ControlValueAccessor, NgControl } from '@angular/forms';
var SwitchComponent = /** @class */ (function () {
    function SwitchComponent(control, elementRef) {
        this.control = control;
        this.elementRef = elementRef;
        this.required = false;
        this.disabled = false;
        this.checked = false;
        this.isDirty = false;
        this.isTouched = false;
        this.onChangeCallback = function () { };
        this.onTouchedCallback = function () { };
        if (this.control) {
            this.control.valueAccessor = this;
        }
    }
    SwitchComponent.prototype.writeValue = function (value) {
        this.checked = value;
    };
    SwitchComponent.prototype.registerOnChange = function (fn) {
        this.onChangeCallback = fn;
    };
    SwitchComponent.prototype.registerOnTouched = function (fn) {
        this.onTouchedCallback = fn;
    };
    SwitchComponent.prototype.setDisabledState = function (disable) {
        this.disabled = disable;
    };
    SwitchComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        Promise.resolve().then(function () {
            if (!_this.id)
                _this.id = Math.random()
                    .toString(36)
                    .substring(2);
            if (!_this.name)
                _this.name = 'switch-' + _this.id;
        });
    };
    SwitchComponent.prototype.onTouched = function () {
        this.onTouchedCallback();
    };
    SwitchComponent.prototype.toggle = function (event) {
        var _a;
        if (this.disabled)
            return;
        var checkedValue = event.target.checked;
        this.onChangeCallback(checkedValue);
        this.onTouched();
        this.isDirty = true;
        this.isTouched = true;
        if ((_a = this.control) === null || _a === void 0 ? void 0 : _a.valueAccessor) {
            var _b = this.control, touched = _b.touched, dirty = _b.dirty, invalid = _b.invalid;
            this.invalid = invalid ? dirty || touched : false;
        }
        else {
            this.invalid = (this.isDirty || this.isTouched) && this.required === true && this.checked === false;
        }
    };
    SwitchComponent.ctorParameters = function () { return [
        { type: NgControl, decorators: [{ type: Self }, { type: Optional }] },
        { type: ElementRef }
    ]; };
    __decorate([
        ViewChild('input')
    ], SwitchComponent.prototype, "input", void 0);
    __decorate([
        Input()
    ], SwitchComponent.prototype, "name", void 0);
    __decorate([
        Input()
    ], SwitchComponent.prototype, "required", void 0);
    __decorate([
        Input()
    ], SwitchComponent.prototype, "id", void 0);
    __decorate([
        Input()
    ], SwitchComponent.prototype, "labelPosition", void 0);
    __decorate([
        HostBinding('class.disabled'),
        Input()
    ], SwitchComponent.prototype, "disabled", void 0);
    __decorate([
        HostBinding('attr.checked'),
        Input()
    ], SwitchComponent.prototype, "checked", void 0);
    SwitchComponent = __decorate([
        Component({
            selector: 'bmw-switch',
            template: "<label\n  [for]=\"id\"\n  class=\"switch\"\n  [ngClass]=\"{ 'label-position-right': labelPosition === 'right', 'label-position-left': labelPosition === 'left' }\"\n>\n  <input\n    type=\"checkbox\"\n    [attr.name]=\"name\"\n    [id]=\"id\"\n    (change)=\"toggle($event)\"\n    [required]=\"required\"\n    [tabindex]=\"disabled ? -1 : 0\"\n    [(ngModel)]=\"checked\"\n    [attr.disabled]=\"disabled ? true : null\"\n    [attr.aria-checked]=\"checked\"\n    #input\n  />\n  <div class=\"track\"></div>\n  <span class=\"label bmw-component-text\"><ng-content></ng-content></span>\n</label>\n",
            styles: [":host{display:block;outline:0;white-space:nowrap;max-width:100%;width:-webkit-fit-content;width:-moz-fit-content;width:fit-content}.switch{position:relative;display:flex;flex:1;align-items:center;cursor:pointer;height:calc(var(--switch__knob__default__height) + 2 * var(--switch__default__border-width))}.switch .track{height:var(--switch__track__default__height);width:var(--switch__track__default__width);flex-shrink:0;position:relative;background-color:var(--switch__track__default__background-color);border:var(--switch__default__border-width) solid var(--switch__track__default__border-color);border-radius:var(--switch__track__default__border-radius);margin-right:var(--spacing-bmw-xs);transition:background-color .1s ease-in-out}.switch .track:before{position:absolute;content:'';display:flex;font-family:iwp;font-size:var(--switch__checked__knob__default__font-size);align-items:center;justify-content:center;height:var(--switch__knob__default__height);width:var(--switch__knob__default__width);top:50%;transform:translateY(-50%);left:calc(var(--switch__default__border-width) * -1);background-color:var(--switch__knob__default__background-color);border:var(--switch__default__border-width) solid var(--switch__knob__default__border-color);border-radius:var(--switch__knob__default__border-radius);transition:transform .1s ease-in-out}.switch .label{white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.switch input{clip:rect(0,0,0,0);margin:1px;width:1px;height:1px;position:absolute;overflow:hidden;outline:0;border:0}.switch input:checked+.track{background-color:var(--switch__checked__track__default__background-color);border-color:var(--switch__checked__track__default__background-color)}.switch input:checked+.track:before{transform:translate(var(--switch__knob__default__travel),-50%);content:'\\e9cd';color:var(--switch__checked__knob__default__color)}.switch input:focus+.track:before{box-shadow:0 0 0 1px var(--color-bmw-basic5),0 0 0 3px var(--color-bmw-highlight)}.switch:hover .track:before{border-color:var(--switch__knob__default__hover__border-color)}.switch.label-position-left .track{order:2;margin-right:0;margin-left:var(--spacing-bmw-xs)}.switch.label-position-left .label,.switch.label-position-right .track{order:1}.switch.label-position-right .label{order:2}:host.disabled .switch{opacity:.5;cursor:no-drop!important}:host.disabled .switch:hover .track:before{border-color:var(--switch__knob__default__border-color)}"]
        }),
        __param(0, Self()), __param(0, Optional())
    ], SwitchComponent);
    return SwitchComponent;
}());
export { SwitchComponent };
