// @bmw-ds/components/esm2015/combo-box/tag/tag.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Output, Input } from '@angular/core';
let TagComponent = class TagComponent {
    constructor() {
        this.focused = false;
        this.disabled = false;
        this.removeEvent = new EventEmitter();
    }
    cancel() {
        if (!this.disabled) {
            this.removeEvent.emit();
        }
    }
};
__decorate([
    Input()
], TagComponent.prototype, "label", void 0);
__decorate([
    Input()
], TagComponent.prototype, "focused", void 0);
__decorate([
    Input()
], TagComponent.prototype, "disabled", void 0);
__decorate([
    Output()
], TagComponent.prototype, "removeEvent", void 0);
TagComponent = __decorate([
    Component({
        selector: 'bmw-tag',
        template: "<div\n  class=\"bmw-tag bmw-component-text\"\n  [class.disabled]=\"disabled\"\n  [attr.aria-label]=\"'Selected option ' + label + '. It can be remove by pressing the backspace or delete key'\"\n  [class.focused]=\"focused\"\n  tabindex=\"-1\"\n>\n  {{ label }}\n  <i class=\"iwp-icon-gen_close\" (click)=\"cancel()\" tabindex=\"-1\"></i>\n</div>\n",
        styles: [":host{display:flex}:host .bmw-tag{display:inline-block;box-sizing:border-box;position:relative;margin:auto;white-space:nowrap;border:1px solid var(--color-bmw-basic3-shade2);padding:0 var(--combo-box__tag__margin-right);padding-right:calc(var(--combo-box__tag__margin-right) + var(--typography__component-text__line-height));background-color:var(--combo-box__dropdown-item__default__background-color)}:host .bmw-tag i{position:absolute;right:var(--combo-box__tag__padding);top:var(--combo-box__tag__padding)}:host .bmw-tag i:focus{outline:0}:host .bmw-tag.focused{border:1px solid!important}:host .bmw-tag:focus{outline:0}:host .bmw-tag.disabled i{cursor:not-allowed;opacity:.5}"]
    })
], TagComponent);
export { TagComponent };
