// @bmw-ds/components/esm5/label/label.component.js
import { __decorate } from "tslib";
import { Component, Input } from '@angular/core';
var LabelComponent = /** @class */ (function () {
    function LabelComponent() {
        this.disabled = false;
    }
    Object.defineProperty(LabelComponent.prototype, "for", {
        set: function (newFor) {
            this.inputId = newFor;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LabelComponent.prototype, "inputId", {
        get: function () {
            return this._inputId;
        },
        set: function (id) {
            this._inputId = id;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        Input()
    ], LabelComponent.prototype, "disabled", void 0);
    __decorate([
        Input()
    ], LabelComponent.prototype, "required", void 0);
    __decorate([
        Input()
    ], LabelComponent.prototype, "for", null);
    LabelComponent = __decorate([
        Component({
            selector: 'bmw-label',
            template: "<label [for]=\"inputId\" [ngClass]=\"{ disabled: disabled, required: required }\" class=\"bmw-regular-text\">\n  <span *ngIf=\"required; else requiredBlock\">\n    <ng-container *ngTemplateOutlet=\"requiredBlock\"></ng-container>\n  </span>\n  <ng-template #requiredBlock><ng-content></ng-content></ng-template>\n</label>\n",
            styles: [":host{display:flex;align-items:center;max-height:var(--inputfield__default__height)}:host label{padding-top:var(--label__default__padding-vertical);padding-bottom:var(--label__default__padding-vertical);padding-left:var(--label__default__padding-left);width:100%;display:block;color:var(--label__default__color);line-height:var(--label__default__line-height);text-overflow:ellipsis;white-space:nowrap;overflow:hidden}:host label.readonly{pointer-events:none}:host label.disabled{cursor:no-drop;color:var(--inputfield__disabled__color)}:host label.required{display:flex}:host label.required span{text-overflow:ellipsis;white-space:nowrap;overflow:hidden}:host label.required:after{content:'\\00a0 *'}"]
        })
    ], LabelComponent);
    return LabelComponent;
}());
export { LabelComponent };
