// @bmw-ds/components/esm2015/navigation-menu/navigation-menu-submenu-item.component.js
import { __decorate } from "tslib";
import { takeUntil } from 'rxjs/operators';
import { NavigationService } from './navigation.service';
import { Component, HostBinding, Input, HostListener, ViewChild } from '@angular/core';
import { Subject } from 'rxjs';
let NavigationMenuSubmenuItemComponent = class NavigationMenuSubmenuItemComponent {
    constructor(_navService) {
        this._navService = _navService;
        this._selected = false;
        this._disabled = false;
        this._destroyed$ = new Subject();
        this.selectedEvent = new Subject();
    }
    set disabled(value) {
        this._disabled = value;
        if (value)
            this._selected = false;
    }
    get disabled() {
        return this._disabled;
    }
    set selected(value) {
        if (this.disabled || value === this._selected)
            return;
        this._selected = value;
    }
    get selected() {
        return this._selected;
    }
    onKeydown(event) {
        const key = event.code || event.keyCode;
        if (['Enter', 13, 'Space', 32].includes(key) && event.target === this.submenuItemWrapper.nativeElement) {
            this.clickEvent(null);
            event.preventDefault();
            event.stopPropagation();
        }
    }
    ngOnInit() {
        if (!this.id) {
            this.id = Math.random()
                .toString(36)
                .substring(2);
        }
        this._navService.selectionChange.pipe(takeUntil(this._destroyed$)).subscribe(item => {
            if (item !== this) {
                this.selected = false;
            }
        });
    }
    ngOnDestroy() {
        this._destroyed$.next();
        this._destroyed$.complete();
    }
    clickEvent($event) {
        $event === null || $event === void 0 ? void 0 : $event.stopPropagation();
        if (this.disabled || this.selected)
            return;
        this._navService.selectItem(this);
        this._selected = true;
        this.selectedEvent.next(this);
    }
};
NavigationMenuSubmenuItemComponent.ctorParameters = () => [
    { type: NavigationService }
];
__decorate([
    HostBinding('attr.id'),
    Input()
], NavigationMenuSubmenuItemComponent.prototype, "id", void 0);
__decorate([
    Input()
], NavigationMenuSubmenuItemComponent.prototype, "disabled", null);
__decorate([
    Input()
], NavigationMenuSubmenuItemComponent.prototype, "selected", null);
__decorate([
    ViewChild('submenuItemWrapper')
], NavigationMenuSubmenuItemComponent.prototype, "submenuItemWrapper", void 0);
__decorate([
    HostListener('keydown', ['$event'])
], NavigationMenuSubmenuItemComponent.prototype, "onKeydown", null);
NavigationMenuSubmenuItemComponent = __decorate([
    Component({
        selector: 'bmw-navigation-menu-submenu-item',
        template: "<div\n  #submenuItemWrapper\n  class=\"submenu-item-wrapper\"\n  [ngClass]=\"{ selected: selected, disabled: disabled }\"\n  (click)=\"clickEvent($event)\"\n  [tabindex]=\"-1\"\n>\n  <ng-content></ng-content>\n</div>\n",
        styles: [".submenu-item-wrapper{display:block;padding:calc(var(--navigation-menu__icon__padding) * 2) 0;cursor:pointer;background-color:var(--navigation-menu__default__submenu__background-color);padding-left:calc(var(--navigation-menu__icon__padding) * 2 + var(--navigation-menu__icon__font-size) + var(--navigation-menu__item__padding-left) + var(--navigation-menu__submenu-item__padding-left));overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.submenu-item-wrapper.selected{background-color:var(--navigation-menu__selected__background-color);color:var(--navigation-menu__selected__color)}.submenu-item-wrapper.selected:hover{background-color:var(--navigation-menu__selected__background-color)}.submenu-item-wrapper:hover{background-color:var(--navigation-menu__hover__background-color)}.submenu-item-wrapper:focus{background-color:var(--navigation-menu__hover__background-color);outline:solid 1px;outline-color:var(--color-bmw-highlight);outline-offset:-1px}.submenu-item-wrapper:focus.selected{background-color:var(--navigation-menu__selected__background-color);outline:solid 1px;outline-color:var(--color-bmw-basic5);outline-offset:-3px}.submenu-item-wrapper.disabled{cursor:no-drop;opacity:.5}"]
    })
], NavigationMenuSubmenuItemComponent);
export { NavigationMenuSubmenuItemComponent };
