// @bmw-ds/components/esm2015/master-detail-list/master-detail-list-row-entry/master-detail-list-row-entry.component.js
import { __decorate } from "tslib";
import { Component, HostBinding, Input } from '@angular/core';
let MasterDetailListRowEntryComponent = class MasterDetailListRowEntryComponent {
    constructor() {
        this.bold = false;
        this.optional = false;
    }
};
__decorate([
    Input()
], MasterDetailListRowEntryComponent.prototype, "label", void 0);
__decorate([
    Input(),
    HostBinding('class.bold')
], MasterDetailListRowEntryComponent.prototype, "bold", void 0);
__decorate([
    Input(),
    HostBinding('class.optional')
], MasterDetailListRowEntryComponent.prototype, "optional", void 0);
MasterDetailListRowEntryComponent = __decorate([
    Component({
        selector: 'bmw-master-detail-list-row-entry',
        template: "<span class=\"label bmw-component-text\" [innerHTML]=\"label\"></span>\n",
        styles: [":host{display:table-cell;height:var(--list-item-entity-height);padding:var(--spacing-bmw-xxs) var(--spacing-bmw-xs);color:var(--list__default__color);border-bottom:none;order:2}:host .label{display:block;overflow:hidden;text-overflow:ellipsis}:host.optional:not(.bold){display:none}:host.bold{font-weight:700;order:1}@media screen and (min-width:37.49rem){:host{padding:var(--spacing-bmw-xs);border-bottom:1px solid var(--list__item__border-bottom-color);order:0}:host.optional:not(.bold){display:table-cell;order:0}:host.bold{order:0}:host:first-child{padding-left:var(--spacing-bmw-s)}:host:last-child{padding-right:var(--spacing-bmw-s)}}:host-context(bmw-master-detail-list.master-detail){border-bottom:none;order:2;padding:var(--spacing-bmw-xxs) var(--spacing-bmw-xs)}:host-context(bmw-master-detail-list.master-detail).optional:not(.bold){display:none}:host-context(bmw-master-detail-list.master-detail).bold{order:1}"]
    })
], MasterDetailListRowEntryComponent);
export { MasterDetailListRowEntryComponent };
