// @bmw-ds/components/esm2015/checkbox/checkbox.component.js
import { __decorate, __param } from "tslib";
import { Component, AfterViewInit, Input, ViewChild, ElementRef, EventEmitter, Output, OnChanges, HostListener, Optional, Self } from '@angular/core';
import { ControlValueAccessor, NgControl } from '@angular/forms';
let CheckboxComponent = class CheckboxComponent {
    constructor(control, elementRef) {
        this.control = control;
        this.elementRef = elementRef;
        this.disabled = false;
        this.checked = "unchecked" /* unchecked */;
        this.required = false;
        this.updateEvent = new EventEmitter();
        this.isDirty = false;
        this.isTouched = false;
        this.onChange = () => { };
        this.onTouched = () => { };
        if (this.control) {
            this.control.valueAccessor = this;
        }
    }
    onKeyDown(event) {
        const key = event.code || event.keyCode;
        switch (key) {
            case 'Space':
            case 32:
                if (!this.disabled) {
                    this.input.nativeElement.indeterminate = false;
                    this.input.nativeElement.checked = !this.input.nativeElement.checked;
                    this.readState();
                    event.preventDefault();
                }
                break;
        }
    }
    writeValue(value) {
        if (value === true) {
            this.checked = "checked" /* checked */;
        }
        else if (value === false) {
            this.checked = "unchecked" /* unchecked */;
        }
        else {
            this.checked = value;
        }
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    ngAfterViewInit() {
        Promise.resolve().then(() => {
            if (!this.id)
                this.id = Math.random()
                    .toString(36)
                    .substring(2);
            if (!this.name)
                this.name = 'checkbox-' + this.id;
            this.checkProperties();
        });
    }
    ngOnChanges() {
        this.checkProperties();
    }
    checkProperties() {
        if (!this.input)
            return;
        if (this.disabled) {
            this.input.nativeElement.disabled = true;
            this.elementRef.nativeElement.classList.add('disabled');
        }
        else {
            this.input.nativeElement.disabled = false;
            this.elementRef.nativeElement.classList.remove('disabled');
        }
        this.updateCheckState();
    }
    readState() {
        var _a;
        if (this.input.nativeElement.indeterminate || this.disabled)
            return;
        const lastState = this.checked;
        this.checked = this.input.nativeElement.checked
            ? "checked" /* checked */
            : this.input.nativeElement.indeterminate
                ? "indeterminate" /* indeterminate */
                : "unchecked" /* unchecked */;
        this.onChange(this.checked === "checked" /* checked */);
        if (lastState !== this.checked) {
            this.updateEvent.emit(this);
        }
        this.isDirty = true;
        this.isTouched = true;
        if (this.onTouched) {
            this.onTouched();
        }
        if ((_a = this.control) === null || _a === void 0 ? void 0 : _a.valueAccessor) {
            const { touched, dirty, invalid } = this.control;
            this.invalid = invalid ? dirty || touched : false;
        }
        else {
            this.invalid =
                (this.isDirty || this.isTouched) && this.required === true && this.checked === "unchecked" /* unchecked */;
        }
    }
    updateCheckState() {
        if (!this.input)
            return;
        if (this.checked === "indeterminate" /* indeterminate */) {
            this.input.nativeElement.indeterminate = true;
            this.input.nativeElement.checked = false;
        }
        else {
            this.input.nativeElement.checked = this.checked === "checked" /* checked */ || false;
            this.input.nativeElement.indeterminate = false;
        }
    }
    stopPropagation(event) {
        event.stopPropagation();
    }
    setDisabledState(disabled) {
        this.disabled = disabled;
        this.checkProperties();
    }
};
CheckboxComponent.ctorParameters = () => [
    { type: NgControl, decorators: [{ type: Self }, { type: Optional }] },
    { type: ElementRef }
];
__decorate([
    ViewChild('input')
], CheckboxComponent.prototype, "input", void 0);
__decorate([
    Input()
], CheckboxComponent.prototype, "name", void 0);
__decorate([
    Input()
], CheckboxComponent.prototype, "disabled", void 0);
__decorate([
    Input()
], CheckboxComponent.prototype, "checked", void 0);
__decorate([
    Input()
], CheckboxComponent.prototype, "required", void 0);
__decorate([
    Input()
], CheckboxComponent.prototype, "id", void 0);
__decorate([
    Output()
], CheckboxComponent.prototype, "updateEvent", void 0);
__decorate([
    HostListener('keydown', ['$event'])
], CheckboxComponent.prototype, "onKeyDown", null);
CheckboxComponent = __decorate([
    Component({
        selector: 'bmw-checkbox',
        template: "<div>\n  <input\n    type=\"checkbox\"\n    [name]=\"name\"\n    [id]=\"id\"\n    [checked]=\"checked === 'checked'\"\n    [required]=\"required\"\n    (change)=\"readState()\"\n    [tabindex]=\"disabled ? -1 : 0\"\n    [attr.aria-checked]=\"checked === 'checked' ? 'true' : checked === 'indeterminate' ? 'mixed' : 'false'\"\n    [class.invalid]=\"invalid\"\n    #input\n  />\n  <label [for]=\"id\" (click)=\"stopPropagation($event)\">\n    <div class=\"bmw-component-text\"><ng-content></ng-content></div>\n\n    <span class=\"checkmark\">\n      <svg\n        viewBox=\"0 0 16 16\"\n        version=\"1.1\"\n        xmlns=\"http://www.w3.org/2000/svg\"\n        xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n      >\n        <title>Checkmark</title>\n        <g stroke=\"none\" stroke-width=\"0\" fill=\"none\" fill-rule=\"evenodd\">\n          <g>\n            <rect x=\"0\" y=\"0\" width=\"16\" height=\"16\" stroke-width=\"1\"></rect>\n            <path\n              class=\"checkmark\"\n              d=\"M7.2734375,11.4296875 L3.32421875,8.1015625 C3.30859367,8.09374996 3.29687504,8.08007822 3.2890625,8.06054688 C3.28124996,8.04101553 3.27734375,8.01953137 3.27734375,7.99609375 C3.27734375,7.98046867 3.27929686,7.96289072 3.28320312,7.94335938 C3.28710939,7.92382803 3.29687492,7.90625008 3.3125,7.890625 L3.8046875,7.29296875 C3.82031258,7.27734367 3.83789053,7.26562504 3.85742188,7.2578125 C3.87695322,7.24999996 3.89843738,7.24609375 3.921875,7.24609375 C3.94531262,7.24609375 3.96484367,7.24804686 3.98046875,7.25195312 C3.99609383,7.25585939 4.01171867,7.26562492 4.02734375,7.28125 L7.2734375,10.0117188 L11.8203125,4.65625 C11.8359376,4.64062492 11.8535155,4.62890629 11.8730469,4.62109375 C11.8925782,4.61328121 11.9140624,4.609375 11.9375,4.609375 C11.9609376,4.609375 11.9804687,4.61132811 11.9960938,4.61523438 C12.0117188,4.61914064 12.0273437,4.62890617 12.0429688,4.64453125 L12.6289062,5.13671875 C12.6601564,5.16796891 12.6777344,5.20507791 12.6816406,5.24804688 C12.6855469,5.29101584 12.6757814,5.32812484 12.6523438,5.359375 L7.49609375,11.40625 C7.48046867,11.4218751 7.46289072,11.4335937 7.44335938,11.4414062 C7.42382803,11.4492188 7.40625008,11.4570312 7.390625,11.4648438 C7.390625,11.4648438 7.38867189,11.4648438 7.38476562,11.4648438 C7.38085936,11.4648438 7.37890625,11.4648438 7.37890625,11.4648438 C7.35546863,11.4648438 7.33593758,11.4609375 7.3203125,11.453125 C7.30468742,11.4453125 7.28906258,11.4375 7.2734375,11.4296875 Z\"\n            ></path>\n            <rect class=\"indeterminate\" x=\"4\" y=\"4\" width=\"8\" height=\"8\"></rect>\n          </g>\n        </g>\n      </svg>\n    </span>\n  </label>\n</div>\n",
        styles: ["input{width:0;height:0;opacity:0;position:absolute;z-index:-1}input:focus+label svg{outline:solid 1px;outline-color:var(--color-bmw-highlight);outline-offset:-1px}div{display:block;position:relative;margin-bottom:var(--checkbox__default__margin-bottom)}div label{position:relative;display:block;padding-left:var(--checkbox__default__padding-left);color:var(--checkbox__default__color);cursor:pointer;width:-webkit-fit-content;width:-moz-fit-content;width:fit-content}div label>span:not(.checkmark){padding-top:calc((var(--checkbox__default__height) - var(--typography__component-text__line-height))/ 2);padding-bottom:calc((var(--checkbox__default__height) - var(--typography__component-text__line-height))/ 2);display:inline-block}div label span.checkmark{top:calc(50% - var(--checkbox__default__height)/ 2);left:0;position:absolute;font-family:iwp;height:var(--checkbox__default__height);width:var(--checkbox__default__height);background-color:var(--checkbox__default_background-color)}div label span.checkmark svg{width:var(--checkbox__default__height);height:var(--checkbox__default__height)}div label span.checkmark svg rect{stroke:var(--checkbox__default__checkbox-color)}div label span.checkmark svg path,div label span.checkmark svg rect.indeterminate{display:none}div label:hover span.checkmark{background-color:var(--checkbox__default__hover__background-color)}div input:checked~label>span.checkmark path.checkmark{display:block;fill:var(--checkbox__default__checkbox-color)}div input:indeterminate~label>span.checkmark rect.indeterminate{display:block;fill:var(--checkbox__default__checkbox-color)}div input[disabled]+label{opacity:.5;pointer-events:none}div input.invalid+label rect{stroke:var(--checkbox__error__color)!important}div input.invalid+label .bmw-component-text{color:var(--checkbox__error__color)}:host.ng-invalid.ng-touched div input+label rect{stroke:var(--checkbox__error__color)!important}:host.ng-invalid.ng-touched div input+label .bmw-component-text{color:var(--checkbox__error__color)}:host.disabled{cursor:no-drop}"]
    }),
    __param(0, Self()), __param(0, Optional())
], CheckboxComponent);
export { CheckboxComponent };
