// @bmw-ds/components/esm5/segmented-button/segmented-button-item/segmented-button-item.component.js
import { __decorate } from "tslib";
import { ToggleButtonComponent } from './../../toggle-button/toggle-button.component';
import { Component, Output, Input, EventEmitter, ViewChild, TemplateRef, ElementRef, OnInit, ChangeDetectorRef } from '@angular/core';
import { FormGroup, FormControl } from '@angular/forms';
var SegmentedButtonItemComponent = /** @class */ (function () {
    function SegmentedButtonItemComponent(el, cd) {
        this.el = el;
        this.cd = cd;
        this.buttonClicked = new EventEmitter();
        this.focused = false;
        this.isInDropdown = false;
        this.itemWidth = 'auto';
        this.native = this.el.nativeElement;
    }
    SegmentedButtonItemComponent.prototype.ngOnInit = function () {
        if (!this.id)
            this.id = Math.random()
                .toString(36)
                .substring(2);
        if (!this.value)
            this.value = this.id;
        var controlObject = {};
        controlObject[this.id] = new FormControl(this.checked);
        this.reactiveForm = new FormGroup(controlObject);
    };
    SegmentedButtonItemComponent.prototype.toggleItemVisibility = function (state) {
        if (typeof state === 'undefined') {
            this.isHidden = !this.isHidden;
        }
        else {
            this.isHidden = state;
        }
    };
    SegmentedButtonItemComponent.prototype.setChecked = function (state) {
        var _this = this;
        setTimeout(function () {
            if (_this.disabled)
                return;
            if (typeof state === 'undefined') {
                _this.checked = !_this.checked;
            }
            else {
                _this.checked = state;
            }
            _this.reactiveForm.controls[_this.id].setValue(_this.checked);
        });
    };
    SegmentedButtonItemComponent.prototype.setTop = function (top) {
        this.fromTop = "calc(" + top * 100 + "% - " + top + "px)";
    };
    SegmentedButtonItemComponent.prototype.setWidth = function (width) {
        this.itemWidth = width + "px";
    };
    SegmentedButtonItemComponent.prototype.moveToDropdown = function () {
        this.isInDropdown = true;
    };
    SegmentedButtonItemComponent.prototype.removeFromDropdown = function () {
        this.itemWidth = 'auto';
        this.isInDropdown = false;
    };
    SegmentedButtonItemComponent.prototype.resetVisibility = function () {
        this.removeFromDropdown();
        this.toggleItemVisibility(false);
    };
    SegmentedButtonItemComponent.prototype.onClick = function (event) {
        if (this.disabled)
            return;
        this.buttonClicked.emit(this);
    };
    SegmentedButtonItemComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        ViewChild('template', { read: TemplateRef })
    ], SegmentedButtonItemComponent.prototype, "sample", void 0);
    __decorate([
        ViewChild(ToggleButtonComponent)
    ], SegmentedButtonItemComponent.prototype, "button", void 0);
    __decorate([
        Input()
    ], SegmentedButtonItemComponent.prototype, "disabled", void 0);
    __decorate([
        Input()
    ], SegmentedButtonItemComponent.prototype, "checked", void 0);
    __decorate([
        Input()
    ], SegmentedButtonItemComponent.prototype, "value", void 0);
    __decorate([
        Input()
    ], SegmentedButtonItemComponent.prototype, "id", void 0);
    __decorate([
        Output()
    ], SegmentedButtonItemComponent.prototype, "buttonClicked", void 0);
    SegmentedButtonItemComponent = __decorate([
        Component({
            selector: 'bmw-segmented-button-item',
            template: "<ng-template #template>\n  <div\n    class=\"button-wrapper\"\n    [ngClass]=\"{\n      overflowed: isInDropdown,\n      hidden: isHidden\n    }\"\n    [style.top]=\"fromTop\"\n    [style.width]=\"itemWidth\"\n    [class.focused]=\"focused\"\n  >\n    <form [formGroup]=\"reactiveForm\">\n      <bmw-toggle-button\n        [disabled]=\"disabled\"\n        [checked]=\"checked && !disabled\"\n        (updateEvent)=\"onClick($event)\"\n        [form]=\"reactiveForm\"\n        [name]=\"id\"\n        [tabIndex]=\"isInDropdown ? -1 : 0\"\n      >\n        <ng-content></ng-content>\n      </bmw-toggle-button>\n    </form>\n  </div>\n</ng-template>\n",
            styles: [".button-wrapper{flex:1 0 auto;max-width:100%;margin-right:-1px}.overflowed{position:absolute;top:100%;right:0;margin-right:0;z-index:var(--segmented-button__item__z-index)}.overflowed.hidden{display:none}.overflowed.focused{z-index:calc(var(--toggle-button__default__z-index) + 1)}.overflowed.focused bmw-toggle-button ::ng-deep label{outline:solid 1px;outline-color:var(--color-bmw-highlight);outline-offset:-1px}.overflowed.focused bmw-toggle-button ::ng-deep label.button--primary{outline:solid 1px;outline-color:var(--color-bmw-basic5);outline-offset:-3px}bmw-toggle-button ::ng-deep label{width:100%;height:100%;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}"]
        })
    ], SegmentedButtonItemComponent);
    return SegmentedButtonItemComponent;
}());
export { SegmentedButtonItemComponent };
