// @bmw-ds/components/esm5/segmented-button/segmented-button.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SegmentedButtonComponent } from './segmented-button.component';
import { ToggleButtonModule } from '../toggle-button/toggle-button.module';
import { ButtonModule } from '../button/button.module';
import { SegmentedButtonItemComponent } from './segmented-button-item/segmented-button-item.component';
var SegmentedButtonModule = /** @class */ (function () {
    function SegmentedButtonModule() {
    }
    SegmentedButtonModule = __decorate([
        NgModule({
            imports: [CommonModule, ToggleButtonModule, ButtonModule],
            declarations: [SegmentedButtonComponent, SegmentedButtonItemComponent],
            entryComponents: [SegmentedButtonItemComponent],
            exports: [SegmentedButtonComponent, SegmentedButtonItemComponent, ToggleButtonModule, ButtonModule]
        })
    ], SegmentedButtonModule);
    return SegmentedButtonModule;
}());
export { SegmentedButtonModule };
export { SegmentedButtonComponent, SegmentedButtonItemComponent };
