// @bmw-ds/components/esm2015/combo-box/combo-box-dropdown-item.component.js
import { __decorate } from "tslib";
import { Component, Input, Output, EventEmitter, HostBinding, ViewChild } from '@angular/core';
let ComboBoxDropdownItemComponent = class ComboBoxDropdownItemComponent {
    constructor() {
        this.disabled = false;
        this.selected = false;
        this.focused = false;
        this.id = Math.random()
            .toString(36)
            .substring(2);
        this.clickEvent = new EventEmitter();
    }
    ngOnChanges() {
        var _a;
        if (this.focused) {
            (_a = this.itemWrapper) === null || _a === void 0 ? void 0 : _a.nativeElement.focus();
        }
    }
    click() {
        if (!this.disabled)
            this.clickEvent.emit();
    }
};
__decorate([
    Input()
], ComboBoxDropdownItemComponent.prototype, "value", void 0);
__decorate([
    Input()
], ComboBoxDropdownItemComponent.prototype, "label", void 0);
__decorate([
    Input()
], ComboBoxDropdownItemComponent.prototype, "disabled", void 0);
__decorate([
    Input()
], ComboBoxDropdownItemComponent.prototype, "selected", void 0);
__decorate([
    Input()
], ComboBoxDropdownItemComponent.prototype, "focused", void 0);
__decorate([
    ViewChild('itemWrapper')
], ComboBoxDropdownItemComponent.prototype, "itemWrapper", void 0);
__decorate([
    HostBinding('attr.id'),
    Input()
], ComboBoxDropdownItemComponent.prototype, "id", void 0);
__decorate([
    Output()
], ComboBoxDropdownItemComponent.prototype, "clickEvent", void 0);
ComboBoxDropdownItemComponent = __decorate([
    Component({
        selector: 'bmw-combo-box-dropdown-item',
        template: "<div\n  #itemWrapper\n  [class.disabled]=\"disabled\"\n  class=\"wrapper\"\n  [attr.aria-label]=\"'Option with label ' + label\"\n  tabindex=\"-1\"\n>\n  <div [class.selected]=\"selected\" [class.focused]=\"focused\" (click)=\"click()\" class=\"dropdown-item\">\n    <div class=\"label bmw-component-text\">{{ label }}</div>\n  </div>\n</div>\n",
        styles: ["div.dropdown-item{width:100%;overflow:hidden;padding:var(--combo-box__default__padding);box-sizing:border-box;position:relative;background-color:var(--combo-box__dropdown-item__default__background-color)}div.dropdown-item div.label{width:100%;display:block;box-sizing:border-box;position:relative;text-overflow:ellipsis;overflow:hidden;white-space:nowrap}div.dropdown-item div.label.bmw-component-text{color:inherit}div.dropdown-item.focused,div.dropdown-item:hover{background-color:var(--combo-box__dropdown-item__hover__background-color)}div.dropdown-item.selected{background-color:var(--combo-box__dropdown-item__selected__background-color)}div.dropdown-item.selected.focused{outline:solid 1px;outline-color:var(--color-bmw-basic5);outline-offset:-3px}div.dropdown-item.selected .label.bmw-component-text{color:var(--combo-box__dropdown-item__selected__color)}.disabled div.dropdown-item{pointer-events:none}.disabled div.dropdown-item .bmw-component-text{opacity:.5}div.wrapper.keyboard-focus .dropdown-item{background-color:var(--combo-box__dropdown-item__hover__background-color)}div.wrapper.disabled{cursor:no-drop}"]
    })
], ComboBoxDropdownItemComponent);
export { ComboBoxDropdownItemComponent };
