// @bmw-ds/components/esm2015/inputfield/input-file/bmw-input-file.directive.js
import { __decorate, __param } from "tslib";
import { Directive, Renderer2, ElementRef, AfterViewInit, HostListener, OnChanges, Input, Injector, SimpleChanges, Optional } from '@angular/core';
import { fromEvent, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { NgControl, FormControl } from '@angular/forms';
let BmwInputFileDirective = class BmwInputFileDirective {
    constructor(injector, _renderer, _el, control) {
        this.injector = injector;
        this._renderer = _renderer;
        this._el = _el;
        this.control = control;
        this.destroy$ = new Subject();
        this.btnLabel = 'Browse';
    }
    onMouseOver() {
        this._renderer.addClass(this.browseBtn, 'active');
    }
    onMouseLeave() {
        this._renderer.removeClass(this.browseBtn, 'active');
    }
    onChange() {
        if (this._el.nativeElement.files.length) {
            const fileName = this._el.nativeElement.files[0].name;
            this._renderer.setAttribute(this.visibleInput, 'value', fileName);
        }
        else {
            this._renderer.setAttribute(this.visibleInput, 'value', '');
        }
        this.checkForRequired();
    }
    checkForRequired() {
        setTimeout(() => {
            const classList = this._el.nativeElement.classList;
            if (this._el.nativeElement.required && classList.contains('ng-invalid') && classList.contains('ng-touched')) {
                this.visibleInput.classList.add('ng-invalid');
                this.visibleInput.classList.add('ng-touched');
            }
            else {
                this.visibleInput.classList.remove('ng-invalid');
                this.visibleInput.classList.remove('ng-touched');
            }
        }, 0);
    }
    _getInputGridClass(el) {
        return el.classList.value.split(' ').find(className => className.includes('bmw__grid__cell'));
    }
    ngOnChanges(changes) {
        if (this.visibleInput && this.browseBtn) {
            this.checkForRequired();
            if (changes.btnLabel) {
                this.updateButtonLabel();
            }
        }
    }
    ngAfterViewInit() {
        if (!this._el.nativeElement.nextElementSibling)
            return;
        const gridClass = this._getInputGridClass(this._el.nativeElement);
        this._renderer.addClass(this._el.nativeElement.nextElementSibling, gridClass);
        this.browseBtn = this._el.nativeElement.nextElementSibling.querySelector('button');
        this.visibleInput = this._el.nativeElement.nextElementSibling.querySelector('input');
        if (this._el.nativeElement.attributes['aria-label']) {
            this.visibleInput.setAttribute('aria-label', this._el.nativeElement.attributes['aria-label'].value);
        }
        const clickOnButton$ = fromEvent(this.browseBtn, 'click').pipe(takeUntil(this.destroy$));
        const clickOnInput$ = fromEvent(this.visibleInput, 'click').pipe(takeUntil(this.destroy$));
        const keydownOnInput$ = fromEvent(this.visibleInput, 'keydown').pipe(takeUntil(this.destroy$));
        const ngControl = this.injector.get(NgControl, null);
        if (ngControl) {
            this.inputControl = ngControl.control;
        }
        clickOnButton$.subscribe(() => {
            var _a;
            this._el.nativeElement.click();
            (_a = this.inputControl) === null || _a === void 0 ? void 0 : _a.markAsTouched();
        });
        clickOnInput$.subscribe(() => {
            var _a;
            this._el.nativeElement.click();
            (_a = this.inputControl) === null || _a === void 0 ? void 0 : _a.markAsTouched();
        });
        keydownOnInput$.subscribe((event) => {
            var _a;
            if (event.code === 'Space' || event.keyCode === 32) {
                this._el.nativeElement.click();
                (_a = this.inputControl) === null || _a === void 0 ? void 0 : _a.markAsTouched();
            }
        });
        if (this.control) {
            this.control.valueChanges.pipe(takeUntil(this.destroy$)).subscribe(value => {
                if (!value) {
                    this.onChange();
                }
            });
        }
        this.checkForRequired();
        this.updateButtonLabel();
    }
    updateButtonLabel() {
        this.browseBtn.querySelector('div').innerText = this.btnLabel;
    }
    destroy() {
        this.destroy$.next();
    }
};
BmwInputFileDirective.ctorParameters = () => [
    { type: Injector },
    { type: Renderer2 },
    { type: ElementRef },
    { type: NgControl, decorators: [{ type: Optional }] }
];
__decorate([
    Input()
], BmwInputFileDirective.prototype, "btnLabel", void 0);
__decorate([
    HostListener('mouseover')
], BmwInputFileDirective.prototype, "onMouseOver", null);
__decorate([
    HostListener('mouseleave')
], BmwInputFileDirective.prototype, "onMouseLeave", null);
__decorate([
    HostListener('change')
], BmwInputFileDirective.prototype, "onChange", null);
BmwInputFileDirective = __decorate([
    Directive({
        selector: 'input[bmwInputFile]'
    }),
    __param(3, Optional())
], BmwInputFileDirective);
export { BmwInputFileDirective };
