// @bmw-ds/components/esm5/progress-bar/progress-bar.component.js
import { __decorate } from "tslib";
import { Component, Input, HostBinding } from '@angular/core';
var ProgressBarComponent = /** @class */ (function () {
    function ProgressBarComponent() {
        this.filled = 0;
        this.duration = 0.5;
        this.filledStyleObject = {
            width: '0%',
            transition: 'width 0.5s'
        };
    }
    ProgressBarComponent.prototype.ngOnInit = function () {
        this.checkProperties();
    };
    ProgressBarComponent.prototype.ngOnChanges = function () {
        this.checkProperties();
    };
    ProgressBarComponent.prototype.checkProperties = function () {
        if (!this.id)
            this.id = Math.random()
                .toString(36)
                .substring(2);
        this.filled = Math.max(0, Math.min(100, this.filled));
        this.filledStyleObject = {
            width: (this.filled ? this.filled.toString() : '0') + '%',
            transition: 'width ' + (this.duration ? this.duration.toString() : '0.5') + 's'
        };
    };
    __decorate([
        HostBinding('attr.id'),
        Input()
    ], ProgressBarComponent.prototype, "id", void 0);
    __decorate([
        Input()
    ], ProgressBarComponent.prototype, "label", void 0);
    __decorate([
        Input()
    ], ProgressBarComponent.prototype, "filled", void 0);
    __decorate([
        Input()
    ], ProgressBarComponent.prototype, "duration", void 0);
    ProgressBarComponent = __decorate([
        Component({
            selector: 'bmw-progress-bar',
            template: "<div class=\"wrapper\" role=\"progressbar\" aria-valuemin=\"0\" aria-valuemax=\"100\" [attr.aria-valuenow]=\"filled\">\n  <div class=\"filled\" [ngStyle]=\"filledStyleObject\"></div>\n  <div class=\"label bmw-regular-text\">\n    {{ label }}\n  </div>\n</div>\n",
            styles: [".wrapper{width:100%;position:relative;background-color:var(--progress-bar__default__background-color)}.wrapper .filled{background-color:var(--progress-bar__filled__background-color);position:absolute;height:100%}.wrapper .label{background-color:rgba(0,0,0,0);position:relative;display:table;margin:auto;line-height:var(--progress-bar__default__line-height)}"]
        })
    ], ProgressBarComponent);
    return ProgressBarComponent;
}());
export { ProgressBarComponent };
