// @bmw-ds/components/esm5/combo-box/combo-box-input.component.js
import { __decorate } from "tslib";
import { Component, HostListener, Output, EventEmitter, HostBinding } from '@angular/core';
/* tslint:disable:component-selector*/
var ComboBoxInputComponent = /** @class */ (function () {
    function ComboBoxInputComponent() {
        this.clickEvent = new EventEmitter();
        this.focusEvent = new EventEmitter();
        this.inputEvent = new EventEmitter();
        this.value = '';
    }
    ComboBoxInputComponent.prototype.onCLick = function () {
        this.clickEvent.emit();
    };
    ComboBoxInputComponent.prototype.onFocus = function () {
        this.focusEvent.emit();
    };
    ComboBoxInputComponent.prototype.onInput = function (event) {
        this.value = event.target.value;
        this.inputEvent.emit();
    };
    __decorate([
        Output()
    ], ComboBoxInputComponent.prototype, "clickEvent", void 0);
    __decorate([
        Output()
    ], ComboBoxInputComponent.prototype, "focusEvent", void 0);
    __decorate([
        Output()
    ], ComboBoxInputComponent.prototype, "inputEvent", void 0);
    __decorate([
        HostBinding('value')
    ], ComboBoxInputComponent.prototype, "value", void 0);
    __decorate([
        HostListener('click')
    ], ComboBoxInputComponent.prototype, "onCLick", null);
    __decorate([
        HostListener('focus')
    ], ComboBoxInputComponent.prototype, "onFocus", null);
    __decorate([
        HostListener('input', ['$event'])
    ], ComboBoxInputComponent.prototype, "onInput", null);
    ComboBoxInputComponent = __decorate([
        Component({
            selector: 'input[bmwComboBoxInput]',
            template: "<ng-content></ng-content>\n",
            styles: ["::ng-deep input[bmwComboBoxInput]{width:100%;min-width:var(--combo-box__inputfield__default__min-width);border:none;border-bottom:1px solid var(--combo-box__inputfield__default__border-bottom-color);border-radius:0;box-sizing:border-box;padding:var(--combo-box__default__padding);padding-bottom:calc(var(--combo-box__default__padding) - 1px);padding-right:calc(var(--typography__component-text__font-size) + 3 * var(--combo-box__default__padding));color:var(--combo-box__inputfield__default__color);font-size:var(--typography__component-text__font-size);line-height:var(--typography__component-text__line-height)}::ng-deep input[bmwComboBoxInput]::-ms-clear{display:none}::ng-deep input[bmwComboBoxInput].filled{padding-right:calc(2 * var(--typography__component-text__font-size) + 4 * var(--combo-box__default__padding))}::ng-deep input[bmwComboBoxInput]::-moz-placeholder{color:var(--combo-box__placeholder__color)}::ng-deep input[bmwComboBoxInput]:-ms-input-placeholder{color:var(--combo-box__placeholder__color)}::ng-deep input[bmwComboBoxInput]::placeholder{color:var(--combo-box__placeholder__color)}::ng-deep input[bmwComboBoxInput].invalid::-moz-placeholder{color:var(--combo-box__error_color)}::ng-deep input[bmwComboBoxInput].invalid:-ms-input-placeholder{color:var(--combo-box__error_color)}::ng-deep input[bmwComboBoxInput].invalid::placeholder{color:var(--combo-box__error_color)}::ng-deep input[bmwComboBoxInput]:focus{outline:0}::ng-deep input[bmwComboBoxInput]:hover{border-bottom-color:var(--combo-box__inputfield__hover__border-bottom-color)}"]
        })
    ], ComboBoxInputComponent);
    return ComboBoxInputComponent;
}());
export { ComboBoxInputComponent };
