// @bmw-ds/components/esm5/combo-box/combo-box.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ComboBoxComponent } from './combo-box.component';
import { ComboBoxInputComponent } from './combo-box-input.component';
import { ComboBoxDropdownItemComponent } from './combo-box-dropdown-item.component';
import { TagComponent } from './tag/tag.component';
import { CheckboxModule } from '../checkbox/checkbox.module';
import { LabelModule } from '../label/label.module';
import { FormsModule } from '@angular/forms';
import { ComboBoxDropdownCheckboxItemComponent } from './combo-box-dropdown-checkbox-item/combo-box-dropdown-checkbox-item.component';
import { ErrorMessageModule } from '../error-message/error-message.module';
var ComboBoxModule = /** @class */ (function () {
    function ComboBoxModule() {
    }
    ComboBoxModule = __decorate([
        NgModule({
            imports: [CommonModule, CheckboxModule, FormsModule, LabelModule, ErrorMessageModule],
            declarations: [
                ComboBoxComponent,
                ComboBoxInputComponent,
                ComboBoxDropdownItemComponent,
                TagComponent,
                ComboBoxDropdownCheckboxItemComponent
            ],
            exports: [ComboBoxComponent, ComboBoxInputComponent, ComboBoxDropdownItemComponent, TagComponent, ErrorMessageModule]
        })
    ], ComboBoxModule);
    return ComboBoxModule;
}());
export { ComboBoxModule };
export { ComboBoxComponent, ComboBoxInputComponent, ComboBoxDropdownItemComponent };
