// @bmw-ds/components/esm2015/label/label.component.js
import { __decorate } from "tslib";
import { Component, Input } from '@angular/core';
let LabelComponent = class LabelComponent {
    constructor() {
        this.disabled = false;
    }
    set for(newFor) {
        this.inputId = newFor;
    }
    set inputId(id) {
        this._inputId = id;
    }
    get inputId() {
        return this._inputId;
    }
};
__decorate([
    Input()
], LabelComponent.prototype, "disabled", void 0);
__decorate([
    Input()
], LabelComponent.prototype, "required", void 0);
__decorate([
    Input()
], LabelComponent.prototype, "for", null);
LabelComponent = __decorate([
    Component({
        selector: 'bmw-label',
        template: "<label [for]=\"inputId\" [ngClass]=\"{ disabled: disabled, required: required }\" class=\"bmw-regular-text\">\n  <span *ngIf=\"required; else requiredBlock\">\n    <ng-container *ngTemplateOutlet=\"requiredBlock\"></ng-container>\n  </span>\n  <ng-template #requiredBlock><ng-content></ng-content></ng-template>\n</label>\n",
        styles: [":host{display:flex;align-items:center;max-height:var(--inputfield__default__height)}:host label{padding-top:var(--label__default__padding-vertical);padding-bottom:var(--label__default__padding-vertical);padding-left:var(--label__default__padding-left);width:100%;display:block;color:var(--label__default__color);line-height:var(--label__default__line-height);text-overflow:ellipsis;white-space:nowrap;overflow:hidden}:host label.readonly{pointer-events:none}:host label.disabled{cursor:no-drop;color:var(--inputfield__disabled__color)}:host label.required{display:flex}:host label.required span{text-overflow:ellipsis;white-space:nowrap;overflow:hidden}:host label.required:after{content:'\\00a0 *'}"]
    })
], LabelComponent);
export { LabelComponent };
