// @bmw-ds/components/esm5/inputfield/bmw-input.directive.js
import { __decorate, __param } from "tslib";
import { InputfieldComponent } from './inputfield.component';
import { BaseInput } from './base-input';
import { Directive, HostBinding, Input, ElementRef, AfterViewInit, Output, EventEmitter, Host, Optional } from '@angular/core';
var SUPPORTED_TYPES = ['text', 'file', 'number', 'password'];
var BmwInputDirective = /** @class */ (function () {
    function BmwInputDirective(_el, _inputField) {
        this._el = _el;
        this._inputField = _inputField;
        this.classBmwInput = true;
        this._type = 'text';
        this.disableToggle = new EventEmitter();
    }
    BmwInputDirective_1 = BmwInputDirective;
    Object.defineProperty(BmwInputDirective.prototype, "isDisabled", {
        // when we bind to attr.disabled we need to set it to 'disabled'
        // or remove it with null because disabled="false" is not the right way
        // to enable the input again, especially the :disabled pseudo selector in css
        // would still be in effect
        get: function () {
            return this.disabled ? 'disabled' : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BmwInputDirective.prototype, "disabled", {
        get: function () {
            return this._inputField && this._inputField.disabled ? true : this._disabled;
        },
        set: function (value) {
            this._disabled = value;
            this.disableToggle.emit(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BmwInputDirective.prototype, "withIcon", {
        set: function (value) {
            if (value) {
                this._el.nativeElement.classList.add('with-icon');
            }
            else {
                this._el.nativeElement.classList.remove('with-icon');
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BmwInputDirective.prototype, "id", {
        get: function () {
            return this._id;
        },
        set: function (value) {
            this._id = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BmwInputDirective.prototype, "type", {
        get: function () {
            return this._type;
        },
        set: function (value) {
            if (SUPPORTED_TYPES.indexOf(value) === -1) {
                throw new Error("Invalid or unsupported type: " + value);
            }
            this._type = value;
        },
        enumerable: true,
        configurable: true
    });
    BmwInputDirective.prototype.ngAfterViewInit = function () {
        var _this = this;
        var list = Array.from(this._el.nativeElement.classList);
        this.placeholder = this._el.nativeElement.placeholder;
        Promise.resolve().then(function () {
            _this.classes = list
                .map(function (item) {
                if (item.match(/bmw__grid__cell.*/)) {
                    return item;
                }
                return null;
            })
                .filter(function (item) { return item; });
        });
    };
    var BmwInputDirective_1;
    BmwInputDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: InputfieldComponent, decorators: [{ type: Optional }, { type: Host }] }
    ]; };
    __decorate([
        HostBinding('class.bmw-input')
    ], BmwInputDirective.prototype, "classBmwInput", void 0);
    __decorate([
        HostBinding('attr.disabled')
    ], BmwInputDirective.prototype, "isDisabled", null);
    __decorate([
        Input()
    ], BmwInputDirective.prototype, "disabled", null);
    __decorate([
        HostBinding('tabindex')
    ], BmwInputDirective.prototype, "tabIndex", void 0);
    __decorate([
        Output()
    ], BmwInputDirective.prototype, "disableToggle", void 0);
    __decorate([
        HostBinding('attr.id'),
        Input()
    ], BmwInputDirective.prototype, "id", null);
    __decorate([
        Input()
    ], BmwInputDirective.prototype, "type", null);
    BmwInputDirective = BmwInputDirective_1 = __decorate([
        Directive({
            selector: 'input[bmwInput], textarea[bmwInput]',
            providers: [
                {
                    provide: BaseInput,
                    useExisting: BmwInputDirective_1
                }
            ]
        }),
        __param(1, Optional()), __param(1, Host())
    ], BmwInputDirective);
    return BmwInputDirective;
}());
export { BmwInputDirective };
