// @bmw-ds/components/esm5/fieldset/fieldset.component.js
import { __decorate } from "tslib";
import { Component, ContentChild, HostBinding, Input, HostListener, ViewChild } from '@angular/core';
import { FieldsetHeaderComponent } from './fieldset-header.component';
var FieldsetComponent = /** @class */ (function () {
    function FieldsetComponent() {
        this._disabled = false;
        this._opened = false;
    }
    Object.defineProperty(FieldsetComponent.prototype, "disabled", {
        get: function () {
            return this._disabled;
        },
        set: function (value) {
            this._disabled = value;
            if (!this.header)
                return;
            this.header.disabled = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FieldsetComponent.prototype, "opened", {
        get: function () {
            return this._opened;
        },
        set: function (value) {
            this._opened = value;
            if (!this.header)
                return;
            this.header.opened = value;
        },
        enumerable: true,
        configurable: true
    });
    FieldsetComponent.prototype.onKeydown = function (event) {
        var key = event.code || event.keyCode;
        switch (key) {
            case 'Enter':
            case 'NumpadEnter':
            case 13:
            case 'Space':
            case 32:
                if (!this.disabled && event.target === this.label.nativeElement) {
                    this.toggle();
                    event.preventDefault();
                    event.stopPropagation();
                }
                break;
        }
    };
    FieldsetComponent.prototype.ngOnInit = function () {
        if (this.header) {
            this.header.opened = this.opened;
        }
        if (!this.id) {
            this.id = Math.random()
                .toString(36)
                .substring(2);
        }
    };
    FieldsetComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        if (this.header) {
            this.header.clickEmit.emit = function () {
                _this.toggle();
            };
        }
    };
    FieldsetComponent.prototype.ngAfterContentInit = function () {
        if (this.header) {
            this.header.opened = this.opened;
            this.header.disabled = this.disabled;
        }
    };
    FieldsetComponent.prototype.toggle = function () {
        if (this.disabled)
            return;
        this.opened = !this.opened;
        if (!this.header)
            return;
        this.header.opened = this.opened;
    };
    __decorate([
        ContentChild(FieldsetHeaderComponent)
    ], FieldsetComponent.prototype, "header", void 0);
    __decorate([
        ViewChild('label')
    ], FieldsetComponent.prototype, "label", void 0);
    __decorate([
        Input()
    ], FieldsetComponent.prototype, "disabled", null);
    __decorate([
        HostBinding('attr.id'),
        Input()
    ], FieldsetComponent.prototype, "id", void 0);
    __decorate([
        Input()
    ], FieldsetComponent.prototype, "opened", null);
    __decorate([
        HostListener('keydown', ['$event'])
    ], FieldsetComponent.prototype, "onKeydown", null);
    FieldsetComponent = __decorate([
        Component({
            selector: 'bmw-fieldset',
            template: "<label #label [class.disabled]=\"disabled\" [tabindex]=\"disabled ? -1 : 0\"\n  ><ng-content select=\"bmw-fieldset-header\"></ng-content\n></label>\n<div [class.disabled]=\"disabled\" class=\"fieldset-container\" *ngIf=\"opened\">\n  <ng-content select=\"bmw-fieldset-container\"></ng-content>\n</div>\n",
            styles: [".fieldset-container{margin-top:var(--fieldset__container__margin-top)}label{display:block}label.disabled{pointer-events:none;opacity:.5}label:focus{outline:solid 1px;outline-color:var(--color-bmw-highlight);outline-offset:-1px}:host[disabled]{cursor:no-drop}"]
        })
    ], FieldsetComponent);
    return FieldsetComponent;
}());
export { FieldsetComponent };
