// @bmw-ds/components/esm5/inline-message/inline-message.component.js
import { __decorate } from "tslib";
import { Component, OnInit, Input, Output, EventEmitter, HostBinding, OnChanges, HostListener, ElementRef } from '@angular/core';
var SUPPORTED_TYPES = ['info', 'success', 'warning', 'error'];
var InlineMessageComponent = /** @class */ (function () {
    function InlineMessageComponent(_el) {
        this._el = _el;
        this._type = '';
        this._messages = [];
        this._showCloseButton = true;
        this.closeEvent = new EventEmitter();
        this.closed = false;
    }
    Object.defineProperty(InlineMessageComponent.prototype, "type", {
        get: function () {
            return this._type;
        },
        set: function (value) {
            if (SUPPORTED_TYPES.indexOf(value) === -1) {
                console.warn("Invalid or unsupported type: " + value);
                this._type = 'info';
            }
            else {
                this._type = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InlineMessageComponent.prototype, "messages", {
        get: function () {
            return this._messages;
        },
        set: function (value) {
            if (typeof value === 'string') {
                this._messages = [value];
            }
            else {
                this._messages = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InlineMessageComponent.prototype, "showCloseButton", {
        get: function () {
            return this._showCloseButton;
        },
        set: function (value) {
            if (value === false) {
                this._showCloseButton = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    InlineMessageComponent.prototype.onKeydown = function (event) {
        var key = event.code || event.keyCode;
        switch (key) {
            case 'Enter':
            case 'NumpadEnter':
            case 13:
            case 'Space':
            case 32:
                if (event.target === this._el.nativeElement.querySelector('.close')) {
                    this.close();
                    event.preventDefault();
                }
                break;
        }
    };
    InlineMessageComponent.prototype.ngOnInit = function () {
        this.checkProperties();
    };
    InlineMessageComponent.prototype.ngOnChanges = function () {
        this.checkProperties();
    };
    InlineMessageComponent.prototype.checkProperties = function () {
        if (!this.id) {
            this.id = Math.random()
                .toString(36)
                .substring(2);
        }
    };
    InlineMessageComponent.prototype.close = function () {
        this.closeEvent.emit();
        this.closed = true;
    };
    InlineMessageComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Input()
    ], InlineMessageComponent.prototype, "type", null);
    __decorate([
        Input()
    ], InlineMessageComponent.prototype, "messages", null);
    __decorate([
        Input()
    ], InlineMessageComponent.prototype, "showCloseButton", null);
    __decorate([
        Input()
    ], InlineMessageComponent.prototype, "groupHeadline", void 0);
    __decorate([
        HostBinding('attr.id'),
        Input()
    ], InlineMessageComponent.prototype, "id", void 0);
    __decorate([
        Output()
    ], InlineMessageComponent.prototype, "closeEvent", void 0);
    __decorate([
        HostListener('keydown', ['$event'])
    ], InlineMessageComponent.prototype, "onKeydown", null);
    InlineMessageComponent = __decorate([
        Component({
            selector: 'bmw-inline-message',
            template: "<div *ngIf=\"!closed\">\n  <div class=\"background\" [ngClass]=\"[type]\">\n    <div class=\"content bmw-component-text \" role=\"alert\" [class.disable-close]=\"!showCloseButton\">\n      <i class=\"iwp-icon- icon\"></i>\n      <p *ngIf=\"messages.length === 0\">\n        <ng-content></ng-content>\n      </p>\n      <p *ngIf=\"messages.length === 1\">\n        {{ messages[0] }}\n      </p>\n      <div *ngIf=\"messages.length > 1\">\n        <p class=\"multimessage__type\">{{ groupHeadline || type }} ({{ messages.length }})</p>\n        <ul>\n          <li *ngFor=\"let message of messages\">{{ message }}</li>\n        </ul>\n      </div>\n      <ng-container *ngIf=\"showCloseButton\">\n        <i class=\"iwp-icon- close\" (click)=\"close()\" tabindex=\"0\" aria-label=\"Close notification\"></i>\n      </ng-container>\n    </div>\n  </div>\n</div>\n",
            styles: ["div.background{border:2px solid var(--inline-message__default__border-color);background-color:#fff;box-sizing:border-box;position:relative}div.background div.content{position:relative;height:100%;background-color:var(--inline-message__default__background-color)}div.background div.content p{padding:calc(var(--inline-message__default__padding) - 2px);padding-left:calc(2 * var(--inline-message__icon__padding) - 2px + var(--inline-message__icon__font-size));margin:0;overflow:hidden;text-overflow:ellipsis}div.background div.content:not(.disable-close) p{padding-right:calc(2 * var(--inline-message__icon__padding) - 2px + var(--inline-message__close__font-size))}div.background div.content i.icon{position:absolute;top:calc(var(--inline-message__default__padding) + var(--typography__component-text__line-height)/ 2 - var(--inline-message__icon__font-size)/ 2 - 2px);left:calc(var(--inline-message__icon__padding) - 2px);font-size:var(--inline-message__icon__font-size);line-height:var(--inline-message__icon__font-size)}div.background div.content i.icon::before{content:'\\e802';color:var(--inline-message__default__border-color)}div.background div.content i.close::after{position:absolute;content:'\\e93e';font-size:var(--inline-message__close__font-size);padding:calc(var(--inline-message__close__padding) - 2px);top:0;right:0;cursor:pointer}div.background div.content i.close:focus::after{outline:solid 1px;outline-color:var(--color-bmw-highlight)}div.background div.content .multimessage__type{text-transform:capitalize;font-weight:700}div.background div.content ul{margin:0;padding-left:calc(2 * var(--inline-message__icon__padding) - 2px + var(--inline-message__icon__font-size));padding-bottom:calc(var(--inline-message__default__padding) - 2px);padding-right:calc(var(--inline-message__default__padding) - 2px);list-style-position:inside}div.background div.content ul li{overflow:hidden;text-overflow:ellipsis}div.background.success{border-color:var(--inline-message__success__border-color)}div.background.success div.content{background-color:var(--inline-message__success__background-color)}div.background.success .content i.icon::before{content:'\\e9d8';color:var(--inline-message__success__border-color)}div.background.warning{border-color:var(--inline-message__warning__border-color)}div.background.warning div.content{background-color:var(--inline-message__warning__background-color)}div.background.warning .content i.icon::before{content:'\\e803';color:var(--inline-message__warning__border-color)}div.background.error{border-color:var(--inline-message__error__border-color)}div.background.error div.content{background-color:var(--inline-message__error__background-color)}div.background.error .content i.icon::before{content:'\\e801';color:var(--inline-message__error__border-color)}"]
        })
    ], InlineMessageComponent);
    return InlineMessageComponent;
}());
export { InlineMessageComponent };
