// @bmw-ds/components/esm5/inputfield/input-stepper/bmw-input-stepper.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, AfterViewInit, Renderer2, OnDestroy, HostListener, HostBinding } from '@angular/core';
import { fromEvent, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
var BmwInputStepperDirective = /** @class */ (function () {
    function BmwInputStepperDirective(_el, _renderer) {
        this._el = _el;
        this._renderer = _renderer;
        this.destroy$ = new Subject();
    }
    BmwInputStepperDirective.prototype.onKeyDown = function (event) {
        var key = event.code || event.keyCode;
        switch (key) {
            case 'ArrowDown':
            case 40:
                this.stepDown();
                event.preventDefault();
                break;
            case 'ArrowUp':
            case 38:
                this.stepUp();
                event.preventDefault();
                break;
        }
    };
    BmwInputStepperDirective.prototype.ngAfterViewInit = function () {
        var _this = this;
        var parent = this._renderer.parentNode(this._el.nativeElement);
        this.increaseBtn = this._createButton('increase');
        this.increaseListener = this._renderer.listen(this.increaseBtn, 'click', this.stepUp.bind(this));
        var keydownOnIncreaseButton$ = fromEvent(this.increaseBtn, 'keydown').pipe(takeUntil(this.destroy$));
        this.decreaseBtn = this._createButton('decrease');
        this.decreaseListener = this._renderer.listen(this.decreaseBtn, 'click', this.stepDown.bind(this));
        var keydownOnDecreaseButton$ = fromEvent(this.decreaseBtn, 'keydown').pipe(takeUntil(this.destroy$));
        keydownOnIncreaseButton$.subscribe(function (event) {
            if (event.code === 'Enter' || event.keyCode === 13) {
                _this.stepUp();
                event.preventDefault();
            }
        });
        keydownOnDecreaseButton$.subscribe(function (event) {
            if (event.code === 'Enter' || event.keyCode === 13) {
                _this.stepDown();
                event.preventDefault();
            }
        });
        var container = this._createContainer();
        this._renderer.appendChild(container, this.increaseBtn);
        this._renderer.appendChild(container, this.decreaseBtn);
        this._renderer.appendChild(parent, container);
    };
    BmwInputStepperDirective.prototype._createContainer = function () {
        var container = this._renderer.createElement('div');
        this._renderer.addClass(container, 'stepper-controls');
        return container;
    };
    BmwInputStepperDirective.prototype._createButton = function (btnName) {
        var spanElement = this._renderer.createElement('span');
        this._renderer.addClass(spanElement, 'stepper-button');
        this._renderer.setAttribute(spanElement, 'tabindex', this._el.nativeElement.disabled ? '-1' : '0');
        this._renderer.addClass(spanElement, btnName);
        return spanElement;
    };
    BmwInputStepperDirective.prototype.ngOnDestroy = function () {
        this.increaseListener();
        this.decreaseListener();
    };
    BmwInputStepperDirective.prototype.stepUp = function () {
        if (!this._el.nativeElement.disabled) {
            this._el.nativeElement.stepUp();
            this._el.nativeElement.dispatchEvent(new Event('input', { bubbles: true }));
        }
    };
    BmwInputStepperDirective.prototype.stepDown = function () {
        if (!this._el.nativeElement.disabled) {
            this._el.nativeElement.stepDown();
            this._el.nativeElement.dispatchEvent(new Event('input', { bubbles: true }));
        }
    };
    BmwInputStepperDirective.prototype.destroy = function () {
        this.destroy$.next();
    };
    BmwInputStepperDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        HostListener('keydown', ['$event'])
    ], BmwInputStepperDirective.prototype, "onKeyDown", null);
    BmwInputStepperDirective = __decorate([
        Directive({
            selector: 'input[bmwInputStepper]'
        })
    ], BmwInputStepperDirective);
    return BmwInputStepperDirective;
}());
export { BmwInputStepperDirective };
