// @bmw-ds/components/esm2015/dropdown/dropdown-item.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, HostBinding, Input, Output } from '@angular/core';
let DropdownItemComponent = class DropdownItemComponent {
    constructor() {
        this.disabled = false;
        this.clicked = new EventEmitter();
        this.keyboardFocus = false;
    }
    ngOnInit() {
        if (!this.id) {
            this.id = Math.random()
                .toString(36)
                .substring(2);
        }
    }
    click() {
        if (!this.disabled)
            this.clicked.emit({ id: this.id, value: this.value });
    }
};
__decorate([
    Input()
], DropdownItemComponent.prototype, "disabled", void 0);
__decorate([
    Input()
], DropdownItemComponent.prototype, "value", void 0);
__decorate([
    Input()
], DropdownItemComponent.prototype, "label", void 0);
__decorate([
    Input()
], DropdownItemComponent.prototype, "id", void 0);
__decorate([
    Output()
], DropdownItemComponent.prototype, "clicked", void 0);
__decorate([
    HostBinding('class.keyboard-focus')
], DropdownItemComponent.prototype, "keyboardFocus", void 0);
DropdownItemComponent = __decorate([
    Component({
        selector: 'bmw-dropdown-item',
        template: "<div [class.disabled]=\"disabled\" class=\"wrapper\">\n  <div [class.selected]=\"selected\" (click)=\"click()\" class=\"dropdown-item\">\n    <div class=\"label bmw-component-text\">{{ label }}</div>\n  </div>\n</div>\n",
        styles: ["div.dropdown-item{width:100%;font-size:var(--dropdown__default__font-size);line-height:var(--dropdown__default__line-height);overflow:hidden;padding-right:var(--dropdown__default__padding-left);padding-top:var(--dropdown__default__padding-top);padding-left:var(--dropdown__default__padding-left);padding-bottom:var(--dropdown__default__padding-bottom);box-sizing:border-box;font-family:Arial;position:relative;background-color:var(--dropdown-item__default__background-color);color:var(--dropdown-item__defualt__color);max-height:calc(2 * var(--dropdown__default__line-height) + var(--dropdown__default__padding-top) + var(--dropdown__default__padding-bottom))}div.dropdown-item div.label{width:100%;display:block;box-sizing:border-box;position:relative;text-overflow:ellipsis;overflow:hidden;white-space:nowrap}div.dropdown-item:hover,div.dropdown-item:hover span::after{background-color:var(--dropdown-item__hover__background-color)}div.dropdown-item.selected{background-color:var(--dropdown-item__selected__background-color)}div.dropdown-item.selected .bmw-component-text{color:var(--dropdown-item__selected__color)}div.dropdown-item.selected span::after{background-color:var(--dropdown-item__selected__background-color)}.disabled div.dropdown-item{pointer-events:none}.disabled div.dropdown-item .bmw-component-text{color:var(--dropdown-item__disabled__color)}div.wrapper.keyboard-focus .dropdown-item,div.wrapper.keyboard-focus .dropdown-item span::after{background-color:var(--dropdown-item__hover__background-color)}div.wrapper.disabled{cursor:no-drop}"]
    })
], DropdownItemComponent);
export { DropdownItemComponent };
