// @bmw-ds/components/esm5/generic-tile/generic-tile.component.js
import { __decorate } from "tslib";
import { Component, OnInit, Input, Output, EventEmitter, OnChanges, HostBinding, HostListener, ElementRef } from '@angular/core';
var GenericTileComponent = /** @class */ (function () {
    function GenericTileComponent(_el) {
        this._el = _el;
        this.selected = false;
        this.clickEvent = new EventEmitter();
        this.clickable = false;
        this.selectable = false;
        this.disabled = false;
    }
    GenericTileComponent.prototype.onKeydown = function (event) {
        var key = event.code || event.keyCode;
        switch (key) {
            case 'Enter':
            case 'NumpadEnter':
            case 13:
            case 'Space':
            case 32:
                if (event.target === this._el.nativeElement.querySelector('.generic-tile')) {
                    this.click();
                    event.preventDefault();
                }
                break;
        }
    };
    GenericTileComponent.prototype.ngOnInit = function () {
        this.checkProperties();
    };
    GenericTileComponent.prototype.ngOnChanges = function () {
        this.checkProperties();
    };
    GenericTileComponent.prototype.checkProperties = function () {
        if (!this.id) {
            this.id = Math.random()
                .toString(36)
                .substring(2);
        }
        this.clickable = this.interaction === 'clickable';
        this.selectable = this.interaction === 'selectable';
    };
    GenericTileComponent.prototype.click = function () {
        if (this.disabled || (!this.clickable && !this.selectable))
            return;
        if (this.selectable) {
            this.selected = !this.selected;
        }
        this.clickEvent.emit({ interaction: this.interaction, selected: this.selected, id: this.id });
    };
    GenericTileComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Input()
    ], GenericTileComponent.prototype, "interaction", void 0);
    __decorate([
        Input()
    ], GenericTileComponent.prototype, "selected", void 0);
    __decorate([
        HostBinding('attr.id'),
        Input()
    ], GenericTileComponent.prototype, "id", void 0);
    __decorate([
        Output()
    ], GenericTileComponent.prototype, "clickEvent", void 0);
    __decorate([
        Input()
    ], GenericTileComponent.prototype, "disabled", void 0);
    __decorate([
        HostListener('keydown', ['$event'])
    ], GenericTileComponent.prototype, "onKeydown", null);
    GenericTileComponent = __decorate([
        Component({
            selector: 'bmw-generic-tile',
            template: "<div\n  class=\"generic-tile\"\n  (click)=\"click()\"\n  [class.clickable]=\"clickable || selectable\"\n  [class.selected]=\"selected && selectable\"\n  [class.disabled]=\"disabled\"\n  [attr.aria-disabled]=\"disabled\"\n  [tabindex]=\"disabled || (!clickable && !selectable) ? -1 : 0\"\n>\n  <ng-content></ng-content>\n</div>\n",
            styles: [":host .generic-tile{box-sizing:border-box;width:100%;padding:var(--generic-tile__default__padding);border:1px solid var(--generic-tile__default__border-color);position:relative;background-color:var(--generic-tile__default__background-color)}:host .generic-tile.clickable{cursor:pointer}:host .generic-tile.clickable.disabled{cursor:no-drop}:host .generic-tile.clickable:hover:not(.disabled){border-color:var(--generic-tile__hover__border-color)}:host .generic-tile.selected{border-color:var(--generic-tile__hover__border-color)}:host .generic-tile.selected::after{content:'\\e9d8';font-family:iwp;font-size:var(--generic-tile__icon__font-size);position:absolute;top:var(--generic-tile__default__padding);right:var(--generic-tile__default__padding);line-height:1;color:var(--generic-tile__selected__icon__color);padding-left:var(--generic-tile__icon__padding);padding-bottom:var(--generic-tile__icon__padding);background-color:var(--generic-tile__default__background-color)}:host .generic-tile.disabled{opacity:var(--generic-tile__disabled__opacity);border-color:var(--generic-tile__disabled__border-color)}:host .generic-tile:focus{outline:0}:host .generic-tile:focus.clickable:not(.disabled){outline:solid 1px!important;outline-color:var(--color-bmw-highlight)!important;outline-offset:-1px}"]
        })
    ], GenericTileComponent);
    return GenericTileComponent;
}());
export { GenericTileComponent };
