// @bmw-ds/components/esm2015/card-tab/card-tab-label.component.js
import { __decorate } from "tslib";
import { Component, ViewChild, AfterViewInit, Output, EventEmitter, ElementRef, Input } from '@angular/core';
let CardTabLabelComponent = class CardTabLabelComponent {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.closable = true;
        this.emitClose = new EventEmitter();
        this.emitClick = new EventEmitter();
        this.destroyed = false;
    }
    ngAfterViewInit() {
        this.closeButton.nativeElement.addEventListener('click', e => {
            this.emitClose.emit(this.for);
            e.stopPropagation();
        });
        this.label.nativeElement.addEventListener('click', e => {
            this.emitClick.emit(this.for);
            e.stopPropagation();
        });
    }
    destroy() {
        this.elementRef.nativeElement.parentElement.removeChild(this.elementRef.nativeElement);
        this.destroyed = true;
    }
};
CardTabLabelComponent.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    ViewChild('close')
], CardTabLabelComponent.prototype, "closeButton", void 0);
__decorate([
    ViewChild('label')
], CardTabLabelComponent.prototype, "label", void 0);
__decorate([
    Input()
], CardTabLabelComponent.prototype, "closable", void 0);
__decorate([
    Input()
], CardTabLabelComponent.prototype, "for", void 0);
__decorate([
    Input()
], CardTabLabelComponent.prototype, "selected", void 0);
__decorate([
    Input()
], CardTabLabelComponent.prototype, "disabled", void 0);
__decorate([
    Output()
], CardTabLabelComponent.prototype, "emitClose", void 0);
__decorate([
    Output()
], CardTabLabelComponent.prototype, "emitClick", void 0);
CardTabLabelComponent = __decorate([
    Component({
        selector: 'bmw-card-tab-label',
        template: "<label #label [class.closable]=\"closable\" [class.disabled]=\"disabled\" [class.selected]=\"selected\">\n  <span class=\"bmw-component-headline-text\"><ng-content></ng-content> </span>\n  <div class=\"close\" #close></div>\n</label>\n",
        styles: ["label{display:block;min-width:var(--card-tabs__label__default__min-width);padding:var(--card-tabs__label__default__padding);padding-bottom:calc(var(--card-tabs__label__default__padding) + 2px);padding-right:calc(var(--card-tabs__label__default__padding) - 2px);margin-right:var(--card-tabs__label__icon__margin-right);text-align:center;color:var(--card-tabs__labels__default__color);height:calc(var(--card-tabs__label__default__height) + 2px);text-overflow:ellipsis;white-space:nowrap;overflow:hidden;bottom:-2px;box-sizing:border-box;cursor:pointer;position:relative}label:focus{outline:0}label.selected{border:2px solid #e5e3dd;border-bottom:none;padding:calc(var(--card-tabs__label__default__padding) - 2px);padding-bottom:calc(var(--card-tabs__label__default__padding) + 2px);background-color:#fff;z-index:var(--card-tabs__menu__z-index)}label.selected .bmw-component-headline-text{color:var(--card-tabs__labels__selected__color)}label.selected div.close::after{padding:calc(var(--card-tabs__label__default__height)/ 2 - (var(--card-tabs__label__with-icon__icon-size)/ 2) - 2px) var(--card-tabs__label__icon__padding)}label.disabled{opacity:.5;pointer-events:none}label.closable{padding-right:calc(2 * var(--card-tabs__label__icon__padding) + var(--card-tabs__label__default__line-height))}label.closable .close::after{content:'\\e93e';font-family:iwp;position:absolute;top:0;right:0;height:var(--card-tabs__label__default__line-height);width:var(--card-tabs__label__default__line-height);color:var(--card-tabs__labels__default__color);padding:var(--card-tabs__label__default__padding) var(--card-tabs__label__icon__padding);padding-bottom:calc(var(--card-tabs__label__default__padding) + 2px)}label:hover .bmw-component-headline-text{color:var(--card-tabs__labels__hover__color)}:host{display:inline-flex;position:relative;max-width:calc(100% - var(--card-tabs__label__default__height))}:host[disabled]{cursor:no-drop}"]
    })
], CardTabLabelComponent);
export { CardTabLabelComponent };
