// @bmw-ds/components/esm2015/toggle-button/toggle-button.component.js
var ToggleButtonComponent_1;
import { __decorate } from "tslib";
import { Component, Input, EventEmitter, Output, forwardRef, HostListener, ViewChild } from '@angular/core';
import { NG_VALUE_ACCESSOR, FormControl, FormBuilder } from '@angular/forms';
let ToggleButtonComponent = ToggleButtonComponent_1 = class ToggleButtonComponent {
    constructor() {
        this.disabled = false;
        this.checked = false;
        this.tabIndex = 0;
        this.updateEvent = new EventEmitter();
        this.primary = false;
        this.secondary = false;
        this.small = false;
        this.onChange = () => { };
        this.onTouched = () => { };
    }
    onKeydown(event) {
        const key = event.code || event.keyCode;
        switch (key) {
            case 'Enter':
            case 'NumpadEnter':
            case 13:
            case 'Space':
            case 32:
                if (!this.disabled) {
                    this.change();
                    this.form.controls[this.name].setValue(this.checked);
                    event.preventDefault();
                }
                break;
        }
    }
    writeValue(value) {
        this.checked = value;
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    ngOnInit() {
        this.generateIfNecessary();
        this.checkProperties();
    }
    ngOnChanges() {
        this.generateIfNecessary();
        this.checkProperties();
    }
    checkProperties() {
        if (this.toggleButton) {
            const properties = this.toggleButton.split(' ');
            if (properties.indexOf('primary') > -1) {
                this.primary = true;
            }
            else if (properties.indexOf('secondary') > -1) {
                this.secondary = true;
            }
            if (properties.indexOf('small') > -1) {
                this.small = true;
            }
        }
        if (this.iconLeft) {
            this.iconRight = undefined;
        }
        if (this.disabled) {
            this.form.controls[this.name].disable();
        }
        else {
            this.form.controls[this.name].enable();
        }
    }
    generateIfNecessary() {
        if (!this.id)
            this.id = Math.random()
                .toString(36)
                .substring(2);
        if (!this.name)
            this.name = 'toggle-button-' + this.id;
        if (!this.form) {
            const control = {};
            control[this.name] = new FormControl(this.checked);
            this.form = new FormBuilder().group(control);
        }
    }
    change() {
        this.checked = !this.checked;
        this.updateEvent.emit(this);
        this.onChange(this.checked);
        if (this.onTouched) {
            this.onTouched();
        }
    }
    focus() {
        this.buttonInput.nativeElement.focus();
    }
    stopPropagation(event) {
        event.stopPropagation();
    }
};
__decorate([
    ViewChild('buttonInput')
], ToggleButtonComponent.prototype, "buttonInput", void 0);
__decorate([
    Input()
], ToggleButtonComponent.prototype, "name", void 0);
__decorate([
    Input()
], ToggleButtonComponent.prototype, "iconRight", void 0);
__decorate([
    Input()
], ToggleButtonComponent.prototype, "iconLeft", void 0);
__decorate([
    Input()
], ToggleButtonComponent.prototype, "disabled", void 0);
__decorate([
    Input()
], ToggleButtonComponent.prototype, "checked", void 0);
__decorate([
    Input()
], ToggleButtonComponent.prototype, "toggleButton", void 0);
__decorate([
    Input()
], ToggleButtonComponent.prototype, "form", void 0);
__decorate([
    Input()
], ToggleButtonComponent.prototype, "id", void 0);
__decorate([
    Input()
], ToggleButtonComponent.prototype, "tabIndex", void 0);
__decorate([
    Output()
], ToggleButtonComponent.prototype, "updateEvent", void 0);
__decorate([
    HostListener('keydown', ['$event'])
], ToggleButtonComponent.prototype, "onKeydown", null);
ToggleButtonComponent = ToggleButtonComponent_1 = __decorate([
    Component({
        selector: 'bmw-toggle-button',
        template: "<div [class.disabled]=\"disabled\" [formGroup]=\"form\">\n  <input\n    #buttonInput\n    type=\"checkbox\"\n    [name]=\"name\"\n    [id]=\"id\"\n    [checked]=\"checked\"\n    [attr.aria-checked]=\"checked\"\n    (change)=\"change()\"\n    [formControlName]=\"name\"\n    [tabindex]=\"disabled ? -1 : tabIndex\"\n  />\n  <label\n    [class.icon-left]=\"iconLeft\"\n    [class.icon-right]=\"iconRight\"\n    [class.button--primary]=\"primary === true\"\n    [class.button--secondary]=\"secondary === true\"\n    [class.button--small]=\"small === true\"\n    [for]=\"id\"\n    (click)=\"stopPropagation($event)\"\n  >\n    <i\n      *ngIf=\"iconLeft\"\n      [ngClass]=\"[\n        iconLeft,\n        'icon-left',\n        small === true ? 'bmw-component-headline-text-small' : 'bmw-component-headline-text'\n      ]\"\n    ></i>\n    <div [ngClass]=\"small === true ? 'bmw-component-headline-text-small' : 'bmw-component-headline-text'\">\n      <ng-content></ng-content>\n    </div>\n    <i\n      *ngIf=\"iconRight\"\n      [ngClass]=\"[\n        iconRight,\n        'icon-right',\n        small === true ? 'bmw-component-headline-text-small' : 'bmw-component-headline-text'\n      ]\"\n    ></i>\n  </label>\n</div>\n",
        providers: [
            {
                provide: NG_VALUE_ACCESSOR,
                useExisting: forwardRef(() => ToggleButtonComponent_1),
                multi: true
            }
        ],
        styles: ["input{width:.1px;height:.1px;opacity:0;position:absolute;z-index:-1}:host{max-width:100%}div label{cursor:pointer;padding:var(--toggle-button__medium__padding);border:solid 1px var(--toggle-button__default__border__color);display:inline-block;box-sizing:border-box;text-align:center;vertical-align:middle;text-transform:uppercase;background-color:var(--toggle-button__default__background-color);position:relative;max-width:100%}div label.icon-left{padding-left:calc(var(--typography__component-headline-text__font-size) + var(--toggle-button__medium__padding) + var(--toggle-button__default__icon__padding))}div label.icon-right{padding-right:calc(var(--typography__component-headline-text__font-size) + var(--toggle-button__medium__padding) + var(--toggle-button__default__icon__padding))}@media all and (-ms-high-contrast:none),(-ms-high-contrast:active){div label{padding-top:calc(var(--toggle-button__medium__padding) + 2px);padding-bottom:calc(var(--toggle-button__medium__padding) - 2px)}div label.icon-left{padding-left:30px}div label.icon-right{padding-right:30px}}div label i{position:absolute}div label i.icon-left{left:var(--toggle-button__medium__padding)}div label i.icon-right{right:var(--toggle-button__medium__padding)}div label i.bmw-component-headline-text{height:var(--typography__component-headline-text__line-height);top:calc(50% - var(--typography__component-headline-text__line-height)/ 2);line-height:var(--typography__component-headline-text__line-height);white-space:nowrap;overflow:hidden;text-overflow:ellipsis}div label i.bmw-component-headline-text-small{height:var(--typography__component-headline-text-small__line-height);top:calc(50% - var(--typography__component-headline-text-small__line-height)/ 2);line-height:var(--typography__component-headline-text-small__line-height)}div label.button--primary{border:solid 1px var(--toggle-button__primary__background-color);background-color:var(--toggle-button__primary__background-color)}div label.button--primary .bmw-component-headline-text,div label.button--primary .bmw-component-headline-text-small{color:var(--toggle-button__primary__color)}div label.button--secondary{border:solid 1px var(--toggle-button__secondary__border__color);background-color:var(--toggle-button__secondary__background-color)}div label:active{border:solid 1px var(--toggle-button__default__active__background-color);background-color:var(--toggle-button__default__active__background-color);color:var(--toggle-button__default__active__color)}div label:active.button--primary{background-color:var(--toggle-button__primary__active__background-color)}div label:active.button--primary .bmw-component-headline-text{color:var(--toggle-button__default__active__color)}div label:active.button--primary .bmw-component-headline-text-small{color:var(--toggle-button__default__active__color)}div label:active.button--secondary{background-color:var(--toggle-button__secondary__active__background-color)}div label:active.button--secondary .bmw-component-headline-text{color:var(--toggle-button__default__active__color)}div label:active.button--secondary .bmw-component-headline-text-small{color:var(--toggle-button__default__active__color)}div label:hover{border:solid 1px var(--toggle-button__default__border__color);background-color:var(--toggle-button__default__hover__background-color)}div label:hover .bmw-component-headline-text,div label:hover .bmw-component-headline-text-small{color:var(--toggle-button__default__hover__color)}div label:hover.button--primary{border:solid 1px var(--toggle-button__primary__hover__background-color);background-color:var(--toggle-button__primary__hover__background-color)}div label:hover.button--primary .bmw-component-headline-text{color:var(--toggle-button__primary__hover__color)}div label:hover.button--primary .bmw-component-headline-text-small{color:var(--toggle-button__primary__hover__color)}div label:hover.button--secondary{border:solid 1px var(--toggle-button__secondary__border__color)}div label:hover.button--secondary .bmw-component-headline-text{color:var(--toggle-button__secondary__hover__color)}div label:hover.button--secondary .bmw-component-headline-text-small{color:var(--toggle-button__secondary__hover__color)}div label.button--small{padding:var(--toggle-button__small__padding);padding-top:var(--toggle-button__small__padding-top);padding-bottom:var(--toggle-button__small__padding-bottom)}div label.button--small.icon-left{padding-left:calc(var(--typography__component-headline-text-small__font-size) + var(--toggle-button__small__padding) + var(--toggle-button__default__icon__padding))}div label.button--small.icon-right{padding-right:calc(var(--typography__component-headline-text-small__font-size) + var(--toggle-button__small__padding) + var(--toggle-button__default__icon__padding))}@media all and (-ms-high-contrast:none),(-ms-high-contrast:active){div label.button--small{padding-top:calc(var(--toggle-button__small__padding-top) + 2px);padding-bottom:calc(var(--toggle-button__small__padding-bottom) - 2px)}div label.button--small.icon-left{padding-left:30px}div label.button--small.icon-right{padding-right:30px}}div label.button--small i.icon-left{left:var(--toggle-button__small__padding)}div label.button--small i.icon-right{right:var(--toggle-button__small__padding)}div input[disabled]+label{opacity:.7;pointer-events:none}div input[aria-checked=true]+label{border:solid 1px var(--toggle-button__default__border__color);background-color:var(--toggle-button__default__selected__background-color);color:var(--toggle-button__default__selected__color)}div input[aria-checked=true]+label .bmw-component-headline-text,div input[aria-checked=true]+label .bmw-component-headline-text-small{color:var(--toggle-button__default__selected__color)}div input[aria-checked=true]+label.button--primary{color:var(--toggle-button__primary__selected__color);border:solid 1px var(--toggle-button__primary__selected__background-color);background-color:var(--toggle-button__primary__selected__background-color)}div input[aria-checked=true]+label.button--primary .bmw-component-headline-text,div input[aria-checked=true]+label.button--primary .bmw-component-headline-text-small{color:var(--toggle-button__primary__selected__color)}div input[aria-checked=true]+label.button--secondary{color:var(--toggle-button__secondary__selected__color);border:solid 1px var(--toggle-button__secondary__border__color);background-color:var(--toggle-button__secondary__selected__background-color)}div input[aria-checked=true]+label.button--secondary .bmw-component-headline-text,div input[aria-checked=true]+label.button--secondary .bmw-component-headline-text-small{color:var(--toggle-button__secondary__selected__color)}div input:focus+label{outline:solid 1px;outline-color:var(--color-bmw-highlight);outline-offset:-1px;z-index:calc(var(--toggle-button__default__z-index) + 1)}div input:focus+label.button--primary{outline:solid 1px;outline-color:var(--color-bmw-basic5);outline-offset:-3px}div.disabled{cursor:no-drop}.bmw-component-headline-text,.bmw-component-headline-text-small{white-space:nowrap;overflow:hidden;text-overflow:ellipsis}"]
    })
], ToggleButtonComponent);
export { ToggleButtonComponent };
