// @bmw-ds/components/esm5/breadcrumb/breadcrumb.component.js
import { __decorate } from "tslib";
import { Component, Input, HostBinding } from '@angular/core';
var BreadcrumbComponent = /** @class */ (function () {
    function BreadcrumbComponent() {
    }
    BreadcrumbComponent.prototype.ngOnInit = function () {
        this.checkProperties();
    };
    BreadcrumbComponent.prototype.ngOnChanges = function () {
        this.checkProperties();
    };
    BreadcrumbComponent.prototype.checkProperties = function () {
        if (!this.id) {
            this.id = Math.random()
                .toString(36)
                .substring(2);
        }
    };
    __decorate([
        HostBinding('attr.id'),
        Input()
    ], BreadcrumbComponent.prototype, "id", void 0);
    BreadcrumbComponent = __decorate([
        Component({
            selector: 'bmw-breadcrumb',
            template: "<nav role=\"navigation\" aria-label=\"Breadcrumbs\">\n  <ng-content></ng-content>\n</nav>\n",
            styles: [":host{display:flex;max-width:100%}"]
        })
    ], BreadcrumbComponent);
    return BreadcrumbComponent;
}());
export { BreadcrumbComponent };
