// @bmw-ds/components/esm5/breadcrumb/breadcrumb-item.component.js
import { __decorate } from "tslib";
import { Component, ViewEncapsulation, Input, HostBinding } from '@angular/core';
var BreadcrumbItemComponent = /** @class */ (function () {
    function BreadcrumbItemComponent() {
    }
    BreadcrumbItemComponent.prototype.ngOnInit = function () {
        this.checkProperties();
    };
    BreadcrumbItemComponent.prototype.ngOnChanges = function () {
        this.checkProperties();
    };
    BreadcrumbItemComponent.prototype.checkProperties = function () {
        if (!this.id) {
            this.id = Math.random()
                .toString(36)
                .substring(2);
        }
    };
    __decorate([
        HostBinding('attr.id'),
        Input()
    ], BreadcrumbItemComponent.prototype, "id", void 0);
    BreadcrumbItemComponent = __decorate([
        Component({
            selector: 'bmw-breadcrumb-item',
            template: "<span class=\"bmw-component-text\">\n  <ng-content select=\"a\"></ng-content>\n</span>\n",
            encapsulation: ViewEncapsulation.None,
            styles: ["bmw-breadcrumb-item{display:var(--breadcrumb__item__default__display);min-width:0;position:relative}bmw-breadcrumb-item span{display:inline-flex;padding-right:var(--breadcrumb__item__label__padding_right);padding-left:var(--breadcrumb__item__label__padding_left);max-width:calc(100% - var(--breadcrumb__item__label__padding_right) - var(--breadcrumb__item__label__padding_left))}bmw-breadcrumb-item a{color:var(--breadcrumb__item__label__default__color);-webkit-text-decoration-line:none;text-decoration-line:none;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}bmw-breadcrumb-item a:hover{-webkit-text-decoration-line:underline;text-decoration-line:underline}bmw-breadcrumb-item a:focus{outline:solid 1px;outline-color:var(--color-bmw-highlight)}bmw-breadcrumb-item:last-child{flex-shrink:0}bmw-breadcrumb-item:last-child span{padding-right:0;max-width:100%}bmw-breadcrumb-item:last-child:after{content:none;padding:0}bmw-breadcrumb-item:last-child a{font-weight:700;color:var(--breadcrumb__item__label__active__color)}bmw-breadcrumb-item:nth-last-child(2){display:inline-flex}bmw-breadcrumb-item:nth-last-child(2):before{content:var(--breadcrumb__item__before__content);left:0}bmw-breadcrumb-item:after,bmw-breadcrumb-item:before{position:absolute;font-family:iwp;color:var(--breadcrumb__item__label__default__color);font-size:var(--breadcrumb__item__arrow__font-size);line-height:var(--breadcrumb__item__arrow__line-height);padding:0 var(--breadcrumb__item__arrow__horizontal-padding);display:inline-flex;-webkit-text-decoration-line:none;text-decoration-line:none}bmw-breadcrumb-item:after{content:var(--breadcrumb__item__after__content);right:0}"]
        })
    ], BreadcrumbItemComponent);
    return BreadcrumbItemComponent;
}());
export { BreadcrumbItemComponent };
