// @bmw-ds/components/esm5/toolbar/toolbar.component.js
import { __decorate, __read, __spread } from "tslib";
import { Component, ContentChildren, HostListener, ViewChild, HostBinding, Input, isDevMode } from '@angular/core';
import { IconButtonComponent } from '../icon-button/icon-button.component';
var ToolbarComponent = /** @class */ (function () {
    function ToolbarComponent() {
        this.visibleIcons = [];
        this.hiddenIcons = [];
        this.opened = false;
        if (isDevMode()) {
            console.warn('Warning - <bmw-toolbar> component is deprecated and will be removed in the near future');
        }
    }
    ToolbarComponent.prototype.listenerFunction = function () {
        var _this = this;
        //wait at least 10mils between re-render
        clearTimeout(this.resizeTimer);
        this.resizeTimer = window.setTimeout(function () {
            _this.onResize(true);
        }, 10);
    };
    ToolbarComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        if (!this.id)
            this.id = Math.random()
                .toString(36)
                .substring(2);
        this.icons.forEach(function (icon, index) {
            _this.visibleIcons.push({
                button: icon,
                index: index
            });
        });
        this.onResize();
    };
    ToolbarComponent.prototype.iconClick = function (icon) {
        icon.clickEvent.emit(icon);
    };
    ToolbarComponent.prototype.onResize = function (showAll) {
        var _this = this;
        if (showAll) {
            this.showAll();
        }
        //wait till DOM re-render
        setTimeout(function () {
            var _iconsWrapperHeight = parseInt(_this.iconsWrapper.nativeElement.getBoundingClientRect().height, 10);
            var _componentHeight = parseInt(_this.wrapper.nativeElement.getBoundingClientRect().height, 10);
            if (_componentHeight < _iconsWrapperHeight) {
                _this.hide();
                //wait till DOM re-render
                setTimeout(function () {
                    _this.onResize();
                });
            }
        });
    };
    ToolbarComponent.prototype.hide = function () {
        var icon = this.visibleIcons.splice(-1, 1);
        this.hiddenIcons = __spread(icon, this.hiddenIcons);
    };
    ToolbarComponent.prototype.showAll = function () {
        this.visibleIcons = __spread(this.visibleIcons, this.hiddenIcons);
        this.hiddenIcons = [];
    };
    ToolbarComponent.prototype.toggleMenu = function () {
        this.opened = !this.opened;
    };
    __decorate([
        ViewChild('wrapper')
    ], ToolbarComponent.prototype, "wrapper", void 0);
    __decorate([
        ViewChild('iconsWrapper')
    ], ToolbarComponent.prototype, "iconsWrapper", void 0);
    __decorate([
        ContentChildren(IconButtonComponent)
    ], ToolbarComponent.prototype, "icons", void 0);
    __decorate([
        HostBinding('attr.id'),
        Input()
    ], ToolbarComponent.prototype, "id", void 0);
    __decorate([
        HostListener('window:resize')
    ], ToolbarComponent.prototype, "listenerFunction", null);
    ToolbarComponent = __decorate([
        Component({
            selector: 'bmw-toolbar',
            template: "<div class=\"toolbar-wrapper\" #wrapper>\n  <div class=\"icons\" #iconsWrapper>\n    <bmw-icon-button\n      *ngFor=\"let icon of visibleIcons\"\n      [iconClass]=\"icon.button.iconClass\"\n      [disabled]=\"icon.button.disabled\"\n      [id]=\"icon.button.id\"\n      [ariaLabel]=\"icon.button.ariaLabel\"\n      (clickEvent)=\"iconClick(icon.button)\"\n    ></bmw-icon-button>\n  </div>\n\n  <div *ngIf=\"hiddenIcons.length > 0\" class=\"overflow\" [class.opened]=\"opened\" (click)=\"toggleMenu()\">\n    <div class=\"overflow-menu\" [class.opened]=\"opened\">\n      <bmw-icon-button\n        *ngFor=\"let icon of hiddenIcons\"\n        [iconClass]=\"icon.button.iconClass\"\n        [disabled]=\"icon.button.disabled\"\n        [id]=\"icon.button.id\"\n        [ariaLabel]=\"icon.button.ariaLabel\"\n        (clickEvent)=\"iconClick(icon.button)\"\n      ></bmw-icon-button>\n    </div>\n  </div>\n</div>\n",
            styles: ["::ng-deep .reversed .toolbar-wrapper{background-color:var(--toolbar__reversed__background-color)!important}.toolbar-wrapper{height:calc(var(--toolbar__default__overflow-icon__width) + 2 * var(--toolbar__default__padding));width:100%;padding:var(--toolbar__default__padding);padding-bottom:calc(var(--toolbar__default__padding) - 1px);padding-right:var(--toolbar__default__padding-right);position:relative;box-sizing:border-box;border-bottom:1px solid var(--toolbar__default__border-bottom-color);background-color:var(--toolbar__default__background-color)}.toolbar-wrapper .icons{display:flex;flex-flow:wrap;width:100%}.toolbar-wrapper .icons bmw-icon-button{padding-right:var(--toolbar__icon__padding-right)}.toolbar-wrapper .overflow{position:absolute;top:var(--toolbar__default__padding);right:var(--toolbar__icon__padding-right);cursor:pointer}.toolbar-wrapper .overflow:after{font-family:iwp;content:'\\ea22';color:var(--toolbar__three-dots__color);font-size:var(--toolbar__default__overflow-icon__width)}.toolbar-wrapper .overflow.opened{background-color:var(--toolbar__three-dots__hover__background-color)}.toolbar-wrapper .overflow .overflow-menu{display:none;position:absolute;top:100%;right:0;background-color:var(--toolbar__overflow-menu__defualt__background-color)}.toolbar-wrapper .overflow .overflow-menu.opened{display:block}.toolbar-wrapper .overflow .overflow-menu ::ng-deep bmw-icon-button{display:inline-block;border:1px solid var(--toolbar__overflow-menu__icon-button__border-color);border-bottom:none}.toolbar-wrapper .overflow .overflow-menu ::ng-deep bmw-icon-button button{padding:calc(var(--icon-button__medium__padding) - 1px);padding-bottom:var(--icon-button__medium__padding)}.toolbar-wrapper .overflow .overflow-menu ::ng-deep bmw-icon-button:last-child{border-bottom:1px solid var(--toolbar__overflow-menu__icon-button__border-color)}.toolbar-wrapper .overflow .overflow-menu ::ng-deep bmw-icon-button:last-child button{padding-bottom:calc(var(--icon-button__medium__padding) - 1px)}"]
        })
    ], ToolbarComponent);
    return ToolbarComponent;
}());
export { ToolbarComponent };
