// @bmw-ds/components/esm5/master-detail-list/master-detail-list-detail/master-detail-list-detail.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
var MasterDetailListDetailComponent = /** @class */ (function () {
    function MasterDetailListDetailComponent() {
        this.closeEvent = new EventEmitter();
        this.hasCloseIcon = true;
        /**
         * Wether to allow overflow from inside content. (default = false)
         */
        this.allowToOverflow = false;
    }
    __decorate([
        Output()
    ], MasterDetailListDetailComponent.prototype, "closeEvent", void 0);
    __decorate([
        Input()
    ], MasterDetailListDetailComponent.prototype, "hasCloseIcon", void 0);
    __decorate([
        Input()
    ], MasterDetailListDetailComponent.prototype, "allowToOverflow", void 0);
    MasterDetailListDetailComponent = __decorate([
        Component({
            selector: 'bmw-master-detail-list-detail',
            template: "<div class=\"master-detail-list-detail\">\n  <div class=\"master-detail-list-detail__header\">\n    <span class=\"bmw-component-headline-text\">\n      <ng-content select=\"[details-header]\"></ng-content>\n    </span>\n    <bmw-icon-button\n      *ngIf=\"hasCloseIcon\"\n      iconClass=\"iwp-icon-gen_close_l\"\n      (clickEvent)=\"closeEvent.emit()\"\n    ></bmw-icon-button>\n  </div>\n</div>\n<div\n  class=\"master-detail-list-detail__content\"\n  [ngClass]=\"{ 'master-detail-list-detail__content--allow-overflow': allowToOverflow }\"\n>\n  <ng-content></ng-content>\n</div>\n",
            styles: [":host{display:flex;flex-direction:column;flex:1 1 auto;background:var(--list-detail__background-color);position:absolute;width:100%;box-shadow:0 6px var(--dialog__default__backdrop-box-shadow-width) -2px rgba(0,0,0,.25)}:host .master-detail-list-detail__header{display:flex;flex:0 0 auto;align-items:center;justify-content:space-between;border-bottom:4px solid var(--list-detail__header__border-color);min-height:calc(var(--icon-button__medium__font-size) + 2 * var(--icon-button__medium__padding))}:host .master-detail-list-detail__header>span{padding:0 var(--spacing-bmw-s)}:host .master-detail-list-detail__content{flex:1 1 auto;overflow:auto;padding:var(--spacing-bmw-s);display:flex;flex-direction:column}:host .master-detail-list-detail__content--allow-overflow{overflow:visible}@media screen and (min-width:37.49rem){:host-context(bmw-master-detail-list.master-detail){position:relative;width:auto;box-shadow:none;margin-left:var(--spacing-bmw-m)}}"]
        })
    ], MasterDetailListDetailComponent);
    return MasterDetailListDetailComponent;
}());
export { MasterDetailListDetailComponent };
